---
title: "RNASeq_Workshop_part3"
author: 
  - name: "Bharat Mishra, Ph.D | bharat26@uab.edu | Senior Bioinformatician"
  - name: "<https://www.uab.edu/cores/ircp/bds>"
date: |
      | Date: `r format(Sys.time(), '%B %d, %Y')`
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Note

This data carpentry workshop materials have been borrowed from several resources including:
1. https://carpentries-incubator.github.io/bioc-rnaseq/index.html
2. https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
3. https://alexslemonade.github.io/refinebio-examples/03-rnaseq/00-intro-to-rnaseq.html
4. https://bioc.ism.ac.jp/packages/3.7/bioc/vignettes/enrichplot/inst/doc/enrichplot.html
5. U-BDS Team.

We thank the curators and maintainers of all of these resources for their wonderful contributions, compiling the best practices, and easy to follow training guides for beginners. 


# Packages loaded globabally

```{r message=FALSE}
# Set the seed so our results are reproducible:
set.seed(2020)

# Required packages

# Mouse annotation package we'll use for gene identifier conversion
library(biomaRt)
library(org.Mm.eg.db)

# pathway analysis
library(gprofiler2)

# We will need them for data handling
library(magrittr)
library(ggrepel)
library(dplyr)
library(tidyverse)
library(readr)

# plotting
library(ggplot2)
library(pheatmap)
library(ComplexHeatmap)
library(RColorBrewer)
library(vidger)
```

### Create `data` and `results` directory

```{r}
dir.create("./data", recursive = TRUE)
dir.create("./results", recursive = TRUE) 
```


# Enrichment Analysis
After we have obtained a list of differentially expressed (DE) genes, the next question naturally to ask is what biological functions these DE genes may affect.

## Gene sets and pathway
A gene set is an unordered collection of genes that are functionally related. A pathway can be interpreted as a gene set by ignoring functional relationships among genes.

### Gene Ontology (GO)
Gene Ontology defines concepts/classes used to describe gene function, and relationships between these concepts. It classifies functions along three aspects:

MF: Molecular Function
    molecular activities of gene products
CC: Cellular Component
    where gene products are active
BP: Biological Process
    pathways and larger processes made up of the activities of multiple gene products
GO terms are organized in a directed acyclic graph, where edges between terms represent parent-child relationship.

### Kyoto Encyclopedia of Genes and Genomes (KEGG)
KEGG is a collection of manually drawn pathway maps representing molecular interaction and reaction networks. These pathways cover a wide range of biochemical processes that can be divided into 7 broad categories:

1. Metabolism
2. Genetic information processing
3. Environmental information processing
4. Cellular processes
5. Organismal systems
6. Human diseases
7. Drug development.

### Other gene sets
GO and KEGG are the most frequently used for functional analysis. They are typically the first choice because of their long-standing curation and availability for a wide range of species.

Other gene sets include but are not limited to Disease Ontology (`DO`), Disease Gene Network (`DisGeNET`), `wikiPathways`, Molecular Signatures Database (`MSigDb`).

## Gene Ontology (GO)

#Input DEG list

```{r}
Transplant_vs_Naive_annotated <- read.csv(file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_annotated_DEGlist.csv", row.names = 1)

normalized_counts_annotated <- read.csv(file="./results/Transplant_vs_Naive/normalized_counts.csv")
```


For the pre-ranked filtered analysis, we will execute GSEA using the standard DEG filters `adj.P.Val <= 0.05, logFC <= -1 | logFC >= 1`.

```{r}
Transplant_vs_Naive_annotated_DEGs <- Transplant_vs_Naive_annotated %>%
    dplyr::filter(padj <= 0.05, log2FoldChange <= -1 | log2FoldChange >= 1) %>% # THIS IS FILTERED ANALYSIS
    dplyr::arrange(dplyr::desc(abs(log2FoldChange))) %>%
    # Filter out the duplicated rows using `dplyr::distinct()`
    dplyr::distinct(external_gene_name, .keep_all = TRUE)
```

Quick look at #of DEGs.
```{r}
nrow(Transplant_vs_Naive_annotated_DEGs)
```

Check for duplicated gene_ids
```{r}
any(duplicated(Transplant_vs_Naive_annotated_DEGs$external_gene_name))
```

# GO by Gprofiler

## set to archive linked to Ensembl 109 version.

```{r}
set_base_url("https://biit.cs.ut.ee/gprofiler_archive3/e109_eg56_p17")

species <- "mmusculus" 

#select data sources, these are our min. standards:

data_sources <- c("GO", "KEGG", "REAC", "MIRNA", "HP", "HPA", "WP")

# although gprofiler2 accepts Ensembl IDs, most users prefer gene ids which can cause redundancy but
# this is to a minority of genes. Thus, we provide results using both inputs and discuss with user
# alternatives when using genes ids

# NOTE: by default we run the following with all DEGs. Another strategy if discussed with PI is to
# split by up/down-regulated.

# You can save URLs for user-friendly way

gost_results <- gost(query = Transplant_vs_Naive_annotated$external_gene_name, 
         organism = species,
         ordered_query = FALSE, 
         multi_query = FALSE,
         significant = TRUE,
         exclude_iea = FALSE, 
         measure_underrepresentation = FALSE, 
         user_threshold = 0.05,
         correction_method = "g_SCS", 
         domain_scope = "custom",
         custom_bg = normalized_counts_annotated$external_gene_name, # genes names from normalized counts
         numeric_ns = "",
         sources = data_sources,
         as_short_link = FALSE,
         evcodes = TRUE)

```

### Explore Gprofiler results

```{r}
gostplot(gost_results, capped = TRUE, interactive = TRUE)
```

```{r}
head(gost_results$result, 5)
```

```{r}
publish_gosttable(gost_results, highlight_terms = gost_results$result[c(1:10),],
                        use_colors = TRUE, 
                        show_columns = c("source", "term_name", "term_size", "intersection_size"),
                        filename = NULL)
```

Write Gprofiler results to csv

```{r}
go_results <- gost_results$result

go_results$parents <- as.character(gost_results$result$parents)
```


```{r}
write.csv(go_results, file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_Gprofiler_padj_fc.csv")
```