---
title: "RNASeq Tertiary Analysis: Part 4"
author: "Bharat Mishra, Ph.D., Austyn Trull, Lara Ianov, Ph.D."
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Packages loaded globabally

```{r message=FALSE}
# Set the seed so our results are reproducible:
set.seed(2020)

# Required packages

# pathway analysis
library(clusterProfiler)
library(msigdbr)

# We will need them for data handling
library(magrittr)
library(ggrepel)
library(dplyr)
library(tidyverse)
library(readr)

# plotting
library(ggplot2)
library(pheatmap)
library(EnhancedVolcano)
library(ComplexUpset)
library(ComplexHeatmap)
library(RColorBrewer)
#library(vidger) # needs fix in container
```

# Gene Set Enrichment Analysis (GSEA)

Gene set enrichment analysis (GSEA) evaluates the associations of a list of DE genes to a collection of pre-defined gene sets, where each gene set has a specific biological meaning.

The definition of a gene set is very flexible and the construction of gene sets is straightforward. In most cases, gene sets are from public databases where huge efforts from scientific curators have already been made to carefully categorize genes into gene sets with clear biological meanings. Nevertheless, gene sets can also be self-defined from individual studies, such as a set of genes in a network module from a co-expression network analysis, or a set of genes that are up-regulated in a certain disease.

There are three key elements of the GSEA method:

1. Calculation of an Enrichment Score (ES)
2. Estimation of Significance Level of ES
3. Adjustment for Multiple Hypothesis Testing

## GSEA and GO enrichment analysis

<!-- TODO: how are they different? -->
<!-- Also discuss how GSEA can be executed on the full dataset regardless of cutoff. How is that different? -->
<!-- [stating that here we will run on DEGs as an example] -->

#Input DEG list


```{r}
Transplant_vs_Naive_annotated <- read.csv(file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_annotated_DEGlist.csv",
                                          row.names = 1)
```


For the pre-ranked filtered analysis, we will execute GSEA using the significance of DEG filters `adj.P.Val <= 0.05`.

```{r}
Transplant_vs_Naive_annotated_DEGs <- Transplant_vs_Naive_annotated %>%
    dplyr::filter(padj <= 0.05) %>% # , log2FoldChange <= -1 | log2FoldChange >= 1 THIS IS NOT FILTERED ANALYSIS
    dplyr::arrange(dplyr::desc(abs(log2FoldChange))) %>%
    # Filter out the duplicated rows using `dplyr::distinct()`
    dplyr::distinct(external_gene_name, .keep_all = TRUE)
```

Quick look at #of DEGs.
```{r}
nrow(Transplant_vs_Naive_annotated_DEGs)
```

Check for duplicated gene_ids
```{r}
any(duplicated(Transplant_vs_Naive_annotated_DEGs$external_gene_name))
```

### Create log2Fold change ranked vector

```{r}
lfc_vector <- Transplant_vs_Naive_annotated_DEGs$log2FoldChange
names(lfc_vector) <- Transplant_vs_Naive_annotated_DEGs$external_gene_name
lfc_vector <- sort(lfc_vector , decreasing = TRUE)

head(lfc_vector)
```

## Get the gene_sets from `MSigDB`

### MSigDB gene sets

Molecular signature database (MSigDB) is a manually curated gene set database. Initially, it was proposed as a supplementary dataset for the original GSEA paper. Later it has been separated out and developed independently. In the first version in 2005, there were only two gene sets collections and in total 843 gene sets. Now in the newest version of MSigDB (v2023.1.Hs), it has grown into nine gene sets collections, covering > 30K gene sets. It provides gene sets on a variety of topics.

MSigDB categorizes gene sets into nine collections where each collection focuses on a specific topic. For some collections, they are additionally split into sub-collections. There are several ways to obtain gene sets from MSigDB. One convenient way is to use the msigdbr package. The advantages is msigdbr supports many other organisms by mapping to orthologs.

Letâ€™s check which organisms are supported and which gene sets collections it provides.

```{r}
msigdbr_species()
```

Look different gene set (gs) collections

```{r}
msigdbr_collections()
```

The first column in the above output is the primary category of gene sets. Some gene sets collections may have sub-collections, and they are shown in the second column.

```r
msigdbr(species, category, subcategory)
```

#### get Hallmark gene set

<!-- TODO: add why would one select Hallmark here -->

```{r}
mm_hallmark_sets <- msigdbr(
  species = "Mus musculus", # Replace with species name relevant to your data
  category = "H"
)

head(mm_hallmark_sets)
```

### Run GSEA

```{r}
# Run GSEA

gsea_enrichment <- GSEA(
  geneList = lfc_vector,
  minGSSize = 25, # minimum gene set
  maxGSSize = 500, # maximum gene set
  pvalueCutoff = 0.05,
  eps = 1e-10, # default
  seed = TRUE,
  pAdjustMethod = "BH", # change the pAdjustMethod as needed
  TERM2GENE = dplyr::select(
    mm_hallmark_sets,
    gs_name,
    gene_symbol)
)
```

### Explore GSEA results

```{r}
# see head excluding the last collumn for simplicity:
head(gsea_enrichment@result)[,1:(ncol(gsea_enrichment@result)-1)]
```

#### Dot plot

```{r fig.height=8, fig.width=8}
dotplot(gsea_enrichment, showCategory=20)
```

** Challenges: Split the dot plot by activated/suppressed **

```{r fig.height=9, fig.width=8}
dotplot(gsea_enrichment, showCategory=20, split=".sign") + facet_grid(.~.sign)
```

```{r}
ridgeplot(gsea_enrichment,
          showCategory = 10,
          fill = "p.adjust",
          core_enrichment = TRUE,
          label_format = 30,
          orderBy = "NES",
          decreasing = FALSE)
```

```{r}
cnetplot(gsea_enrichment, 
         showCategory = 2, 
         foldChange=lfc_vector, 
         colorEdge = TRUE,
         cex_category = 1,
         cex_gene = 1,
         cex_label_category = 0.5,
         cex_label_gene = 0.5,
         circular = FALSE)
```

** Challenges: make the circular network **

```{r fig.height=7, fig.width=10}
cnetplot(gsea_enrichment, 
         showCategory = 2, 
         foldChange=lfc_vector, 
         colorEdge = TRUE,
         cex_category = 0.5,
         cex_gene = 0.5,
         cex_label_category = 0.5,
         cex_label_gene = 0.5,
         circular = TRUE)
```


```{r fig.height=4, fig.width=25}
heatplot(gsea_enrichment, showCategory = 2, foldChange=lfc_vector)
```

#### Most Positive NES

```{r}
gsea_result_df <- data.frame(gsea_enrichment@result)

gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_max(NES, n = 3)
```

```{r fig.height=5, fig.width=6}
most_positive_nes_plot <- enrichplot::gseaplot(
  gsea_enrichment,
  geneSetID = "HALLMARK_G2M_CHECKPOINT",
  title = "HALLMARK_G2M_CHECKPOINT",
  color.line = "#0d76ff"
)
most_positive_nes_plot
```

### Most Negative NES

```{r}
gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_min(NES, n = 3)
```

```{r fig.height=5, fig.width=6}
most_negative_nes_plot <- enrichplot::gseaplot(
  gsea_enrichment,
  geneSetID = "HALLMARK_ALLOGRAFT_REJECTION",
  title = "HALLMARK_ALLOGRAFT_REJECTION",
  color.line = "#0d76ff"
)
most_negative_nes_plot
```

Write GSEA results to csv

```{r}
write.csv(gsea_enrichment@result, file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_GSEA_hallmark_padj_fc.csv")
```