---
title: "RNASeq Tertiary Analysis: Part 3"
author: "Bharat Mishra, Ph.D., Austyn Trull, Lara Ianov, Ph.D."
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Packages loaded globabally

```{r message=FALSE}
# Set the seed so our results are reproducible:
set.seed(2020)

# Required packages

# Mouse annotation package we'll use for gene identifier conversion
library(biomaRt)
library(org.Mm.eg.db)

# pathway analysis
library(gprofiler2)

# We will need them for data handling
library(magrittr)
library(ggrepel)
library(dplyr)
library(tidyverse)
library(readr)

# plotting
library(ggplot2)
library(pheatmap)
library(ComplexHeatmap)
library(RColorBrewer)
#library(vidger) # needs fix in container
```

# Enrichment Analysis
After we have obtained a list of DEG genes, the next question naturally to ask is what biological functions these DE genes may affect.

## Gene sets and pathway
A gene set is an unordered collection of genes that are functionally related. A pathway can be interpreted as a gene set by ignoring functional relationships among genes.

### Gene Ontology (GO)
Gene Ontology defines concepts/classes used to describe gene function, and relationships between these concepts. It classifies functions along three aspects:

MF: Molecular Function
    molecular activities of gene products
CC: Cellular Component
    where gene products are active
BP: Biological Process
    pathways and larger processes made up of the activities of multiple gene products
GO terms are organized in a directed acyclic graph, where edges between terms represent parent-child relationship.

### Kyoto Encyclopedia of Genes and Genomes (KEGG)
KEGG is a collection of manually drawn pathway maps representing molecular interaction and reaction networks. These pathways cover a wide range of biochemical processes that can be divided into 7 broad categories:

1. Metabolism
2. Genetic information processing
3. Environmental information processing
4. Cellular processes
5. Organismal systems
6. Human diseases
7. Drug development.

### Other gene sets
GO and KEGG are the most frequently used for functional analysis. They are typically the first choice because of their long-standing curation and availability for a wide range of species.

Other gene sets include but are not limited to Disease Ontology (`DO`), Disease Gene Network (`DisGeNET`), `wikiPathways`, Molecular Signatures Database (`MSigDb`).

## Input DEG list and DEG filtering

```{r}
Transplant_vs_Naive_annotated <- read.csv(file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_annotated_DEGlist.csv",
                                          row.names = 1)

normalized_counts_annotated <- read.csv(file="./results/Transplant_vs_Naive/normalized_counts.csv")
```


For the pre-ranked filtered analysis, we will execute GSEA using the standard DEG filters `adj.P.Val <= 0.05, logFC <= -1 | logFC >= 1`.

```{r}
Transplant_vs_Naive_annotated_DEGs <- Transplant_vs_Naive_annotated %>%
  dplyr::filter(padj <= 0.05, log2FoldChange <= -1 | log2FoldChange >= 1) %>% # THIS IS FILTERED ANALYSIS
  dplyr::arrange(dplyr::desc(abs(log2FoldChange))) %>%
  # Filter out the duplicated rows using `dplyr::distinct()`
  dplyr::distinct(external_gene_name, .keep_all = TRUE)
```

Quick look at the number of DEGs.
```{r}
nrow(Transplant_vs_Naive_annotated_DEGs)
```

Check for duplicated gene_ids
```{r}
any(duplicated(Transplant_vs_Naive_annotated_DEGs$external_gene_name))
```

## GO by `gprofiler2`

Similarly to prior examples, `gprofiler2` also maintains archived versions based
on genome versions, specifically Ensembl. Thus, here we also set the archive
to Ensembl 109 version.

Further, note that although `gprofiler2` accepts Ensembl IDs, 
most biologists/scientist prefer gene names as they are more interpretable.
One caveat, is that the use of gene names can can lead to redundancies for cases
where a name corresponds to more than one id (e.g: haplotypes). However, for the
most part it is acceptable to use gene names. The example below is executed
with gene names.

Lastly, note we set the `custom_bg` to the gene names present in our normalized
counts which is derived from the raw data that was pre-filtered. This is important
because ... <!-- TODO: briefly touch on why adding a background is important -->

```{r}
set_base_url("https://biit.cs.ut.ee/gprofiler_archive3/e109_eg56_p17")

species <- "mmusculus" 

#select data sources, these are our min. standards:

data_sources <- c("GO", "KEGG", "REAC", "MIRNA", "HP", "HPA", "WP")

gost_results <- gost(query = Transplant_vs_Naive_annotated_DEGs$external_gene_name, 
                     organism = species,
                     ordered_query = FALSE, 
                     multi_query = FALSE,
                     significant = TRUE,
                     exclude_iea = FALSE, 
                     measure_underrepresentation = FALSE, 
                     user_threshold = 0.05,
                     correction_method = "g_SCS", 
                     domain_scope = "custom",
                     custom_bg = normalized_counts_annotated$external_gene_name, # genes names from normalized counts
                     numeric_ns = "",
                     sources = data_sources,
                     as_short_link = FALSE,
                     evcodes = TRUE)
```

### Explore `gprofiler2` results

```{r}
gostplot(gost_results, capped = TRUE, interactive = TRUE)
```

```{r}
head(gost_results$result, 5)
```

```{r}
publish_gosttable(gost_results, highlight_terms = gost_results$result[c(1:10),],
                  use_colors = TRUE, 
                  show_columns = c("source", "term_name", "term_size", "intersection_size"),
                  filename = NULL)
```

Write `gprofiler2` results to csv

```{r}
go_results <- gost_results$result

go_results$parents <- as.character(gost_results$result$parents)
```


```{r}
write.csv(go_results,
          file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_Gprofiler_padj_fc.csv",
          row.names = FALSE)
```

<!-- TODO: add challenge on enabling running `gost` with URL output. Go through output of this in class -->