---
title: "RNASeq_Workshop_part4"
author: 
  - name: "Bharat Mishra, Ph.D | bharat26@uab.edu | Senior Bioinformatician"
  - name: "<https://www.uab.edu/cores/ircp/bds>"
date: |
      | Date: `r format(Sys.time(), '%B %d, %Y')`
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Note

This data carpentry workshop materials have been borrowed from several resources including:
1. https://carpentries-incubator.github.io/bioc-rnaseq/index.html
2. https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
3. https://alexslemonade.github.io/refinebio-examples/03-rnaseq/00-intro-to-rnaseq.html
4. https://bioc.ism.ac.jp/packages/3.7/bioc/vignettes/enrichplot/inst/doc/enrichplot.html
5. U-BDS Team.

We thank the curators and maintainers of all of these resources for their wonderful contributions, compiling the best practices, and easy to follow training guides for beginners. 

## Install these packages
```{r, eval=FALSE}
install.packages(c("BiocManager", "remotes"))
BiocManager::install(c("tidyverse", "SummarizedExperiment",
                       "ExploreModelMatrix", "AnnotationDbi", "org.Hs.eg.db", 
                       "org.Mm.eg.db", "csoneson/ConfoundingExplorer",
                       "DESeq2", "vsn", "ComplexHeatmap", "hgu95av2.db",
                       "RColorBrewer", "hexbin", "cowplot", "iSEE",
                       "clusterProfiler", "enrichplot", "kableExtra",
                       "msigdbr", "gplots", "ggplot2", "simplifyEnrichment",
                       "apeglm", "microbenchmark", "Biostrings",
                       "SingleCellExperiment"))

BiocManager::install(c('edgeR','limma','tximport', 'biomaRt',
                       'Glimma', 'GeneOverlap',
                       'EnhancedVolcano'))

BiocManager::install(c('clusterProfiler', 'ComplexUpset', 'openxlsx','ggrepel','ggpubr'))
```

# Packages loaded globabally

```{r message=FALSE}
# Set the seed so our results are reproducible:
set.seed(2020)

# Required packages
library(tximport)
library(limma) 
library(edgeR)
library(DESeq2)
library(Glimma)
library(vsn)

# Mouse annotation package we'll use for gene identifier conversion
library(biomaRt)
library(org.Mm.eg.db)

# pathway analysis
library(clusterProfiler)
library(msigdbr)

# We will need them for data handling
library(magrittr)
library(ggrepel)
library(dplyr)
library(tidyverse)
library(readr)

# plotting
library(ggplot2)
library(pheatmap)
library(EnhancedVolcano)
library(ComplexUpset)
library(ComplexHeatmap)
library(RColorBrewer)
library(vidger)
```

### Create `data` and `results` directory

```{r}
dir.create("./data", recursive = TRUE)
dir.create("./results", recursive = TRUE) 
```


# Enrichment Analysis
After we have obtained a list of differentially expressed (DE) genes, the next question naturally to ask is what biological functions these DE genes may affect.

## Gene sets and pathway
A gene set is an unordered collection of genes that are functionally related. A pathway can be interpreted as a gene set by ignoring functional relationships among genes.

### Gene Ontology (GO)
Gene Ontology defines concepts/classes used to describe gene function, and relationships between these concepts. It classifies functions along three aspects:

MF: Molecular Function
    molecular activities of gene products
CC: Cellular Component
    where gene products are active
BP: Biological Process
    pathways and larger processes made up of the activities of multiple gene products
GO terms are organized in a directed acyclic graph, where edges between terms represent parent-child relationship.

### Kyoto Encyclopedia of Genes and Genomes (KEGG)
KEGG is a collection of manually drawn pathway maps representing molecular interaction and reaction networks. These pathways cover a wide range of biochemical processes that can be divided into 7 broad categories:

1. Metabolism
2. Genetic information processing
3. Environmental information processing
4. Cellular processes
5. Organismal systems
6. Human diseases
7. Drug development.

### Other gene sets
GO and KEGG are the most frequently used for functional analysis. They are typically the first choice because of their long-standing curation and availability for a wide range of species.

Other gene sets include but are not limited to Disease Ontology (`DO`), Disease Gene Network (`DisGeNET`), `wikiPathways`, Molecular Signatures Database (`MSigDb`).


## Gene Set Enrichment Analysis (GSEA)

Gene set enrichment analysis (GSEA) evaluates the associations of a list of DE genes to a collection of pre-defined gene sets, where each gene set has a specific biological meaning.

The definition of a gene set is very flexible and the construction of gene sets is straightforward. In most cases, gene sets are from public databases where huge efforts from scientific curators have already been made to carefully categorize genes into gene sets with clear biological meanings. Nevertheless, gene sets can also be self-defined from individual studies, such as a set of genes in a network module from a co-expression network analysis, or a set of genes that are up-regulated in a certain disease.

There are three key elements of the GSEA method:

1. Calculation of an Enrichment Score (ES)
2. Estimation of Significance Level of ES
3. Adjustment for Multiple Hypothesis Testing

#Input DEG list


```{r}
Transplant_vs_Naive_annotated <- read.csv(file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_annotated_DEGlist.csv", row.names = 1)
```


For the pre-ranked filtered analysis, we will execute GSEA using the standard DEG filters `adj.P.Val <= 0.05, logFC <= -1 | logFC >= 1`.

```{r}
Transplant_vs_Naive_annotated_DEGs <- Transplant_vs_Naive_annotated %>%
    dplyr::filter(padj <= 0.05, log2FoldChange <= -1 | log2FoldChange >= 1) %>% # THIS IS FILTERED ANALYSIS
    dplyr::arrange(dplyr::desc(abs(log2FoldChange))) %>%
    # Filter out the duplicated rows using `dplyr::distinct()`
    dplyr::distinct(external_gene_name, .keep_all = TRUE)
```

Quick look at #of DEGs.
```{r}
nrow(Transplant_vs_Naive_annotated_DEGs)
```

Check for duplicated gene_ids
```{r}
any(duplicated(Transplant_vs_Naive_annotated_DEGs$external_gene_name))
```

### Create log2Fold change ranked vector

```{r}
lfc_vector <- Transplant_vs_Naive_annotated_DEGs$log2FoldChange
names(lfc_vector) <- Transplant_vs_Naive_annotated_DEGs$external_gene_name
lfc_vector <- sort(lfc_vector , decreasing = TRUE)

head(lfc_vector)
```

## Get the gene_sets from `MSigDB`

### MSigDB gene sets

Molecular signature database (MSigDB) is a manually curated gene set database. Initially, it was proposed as a supplementary dataset for the original GSEA paper. Later it has been separated out and developed independently. In the first version in 2005, there were only two gene sets collections and in total 843 gene sets. Now in the newest version of MSigDB (v2023.1.Hs), it has grown into nine gene sets collections, covering > 30K gene sets. It provides gene sets on a variety of topics.

MSigDB categorizes gene sets into nine collections where each collection focuses on a specific topic. For some collections, they are additionally split into sub-collections. There are several ways to obtain gene sets from MSigDB. One convenient way is to use the msigdbr package. The advantages is msigdbr supports many other organisms by mapping to orthologs.

Letâ€™s check which organisms are supported and which gene sets collections it provides.

```{r}
msigdbr_species()
```

Look different gene set (gs) collections

```{r}
msigdbr_collections()
```

The first column in the above output is the primary category of gene sets. Some gene sets collections may have sub-collections, and they are shown in the second column.

```r
msigdbr(species, category, subcategory)
```

#### get Hallmark gene set

```{r}
mm_hallmark_sets <- msigdbr(
  species = "Mus musculus", # Replace with species name relevant to your data
  category = "H"
)

head(mm_hallmark_sets)
```

### Run GSEA

```{r}
# Run GSEA

gsea_enrichment <- GSEA(
      geneList = lfc_vector,
      minGSSize = 25,        #minimum gene set
      maxGSSize = 500,       #maximum gene set
      pvalueCutoff = 0.05,
      eps = 1e-10, # default
      seed = TRUE,
      pAdjustMethod = "BH", #change the pAdjustMethod as needed
      TERM2GENE = dplyr::select(
        mm_hallmark_sets,
        gs_name,
        gene_symbol)
      )
```

### Explore GSEA results

```{r}
head(gsea_enrichment@result)
```

#### Dot plot

```{r}
dotplot(gsea_enrichment, showCategory=20)
```

** Challenges: Split the dot plot by activated/suppressed **

```{r}
dotplot(gsea_enrichment, showCategory=20, split=".sign") + facet_grid(.~.sign)
```

```{r}
ridgeplot(gsea_enrichment,
          showCategory = 10,
          fill = "p.adjust",
          core_enrichment = TRUE,
          label_format = 30,
          orderBy = "NES",
          decreasing = FALSE)
```

```{r}
cnetplot(gsea_enrichment, 
         showCategory = 2, 
         foldChange=lfc_vector, 
         colorEdge = TRUE,
         cex_category = 1,
         cex_gene = 1,
         cex_label_category = 0.5,
         cex_label_gene = 0.5,
         circular = FALSE)
```

** Challenges: make the circular network **

```{r fig.height=7, fig.width=10}
cnetplot(gsea_enrichment, 
         showCategory = 2, 
         foldChange=lfc_vector, 
         colorEdge = TRUE,
         cex_category = 0.5,
         cex_gene = 0.5,
         cex_label_category = 0.5,
         cex_label_gene = 0.5,
         circular = TRUE)
```


```{r fig.height=4, fig.width=25}
heatplot(gsea_enrichment, foldChange=lfc_vector)
```

#### Most Positive NES

```{r}
gsea_result_df <- data.frame(gsea_enrichment@result)

gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_max(NES, n = 3)
```

```{r fig.height=5, fig.width=6}
most_positive_nes_plot <- enrichplot::gseaplot(
  gsea_enrichment,
  geneSetID = "HALLMARK_G2M_CHECKPOINT",
  title = "HALLMARK_G2M_CHECKPOINT",
  color.line = "#0d76ff"
)
most_positive_nes_plot
```

### Most Negative NES

```{r}
gsea_result_df %>%
  # This returns the 3 rows with the largest NES values
  dplyr::slice_min(NES, n = 3)
```

```{r fig.height=5, fig.width=6}
most_negative_nes_plot <- enrichplot::gseaplot(
  gsea_enrichment,
  geneSetID = "HALLMARK_ALLOGRAFT_REJECTION",
  title = "HALLMARK_ALLOGRAFT_REJECTION",
  color.line = "#0d76ff"
)
most_negative_nes_plot
```

Write GSEA results to csv

```{r}
write.csv(gsea_enrichment@result, file = "./results/Transplant_vs_Naive/Transplant_vs_Naive_GSEA_hallmark_padj_fc.csv")
```