<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Bharat Mishra, Ph.D., Austyn Trull, Lara Ianov, Ph.D." />


<title>RNA-Seq Tertiary Analysis: Part 3</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet" />
<script src="site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
<script src="site_libs/plotly-binding-4.10.4/plotly.js"></script>
<script src="site_libs/typedarray-0.1/typedarray.min.js"></script>
<link href="site_libs/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet" />
<script src="site_libs/crosstalk-1.2.1/js/crosstalk.min.js"></script>
<link href="site_libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
<script src="site_libs/plotly-main-2.11.1/plotly-latest.min.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">U-BDS bulk RNA-seq workshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="installation.html">
    <span class="fa fa-computer"></span>
     
    Install instructions
  </a>
</li>
<li>
  <a href="secondary_analysis.html">
    <span class="fa fa-database"></span>
     
    Secondary analysis
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-chart-simple"></span>
     
    Tertiary analysis
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="tertiary_analysis_Part1.html">Part 1</a>
    </li>
    <li>
      <a href="tertiary_analysis_Part2.html">Part 2</a>
    </li>
    <li>
      <a href="tertiary_analysis_Part3.html">Part 3</a>
    </li>
    <li>
      <a href="tertiary_analysis_Part4.html">Part 4</a>
    </li>
  </ul>
</li>
<li>
  <a href="additional_resources.html">
    <span class="fa fa-book"></span>
     
    Additional resources
  </a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">RNA-Seq Tertiary Analysis: Part 3</h1>
<h4 class="author">Bharat Mishra, Ph.D., Austyn Trull, Lara Ianov,
Ph.D.</h4>

</div>


<div id="packages-loaded-globabally" class="section level1">
<h1>Packages loaded globabally</h1>
<pre class="r"><code># Set the seed so our results are reproducible:
set.seed(2020)

# Required packages

# pathway analysis
library(gprofiler2)

# We will need them for data handling
library(magrittr)
library(ggrepel)
library(dplyr)
library(tidyverse)
library(readr)

# plotting
library(ggplot2)
library(ComplexHeatmap)
library(RColorBrewer)</code></pre>
</div>
<div id="gene-ontology-go" class="section level1">
<h1>Gene Ontology (GO)</h1>
<p>Gene Ontology (GO) enrichment analysis is a computational method used
to identify which biological, cellular, or molecular processes are
over-represented (or under-represented) in a given list of genes. It
leverages the structured vocabulary of the Gene Ontology to interpret
the functions of genes within a larger biological context.</p>
<p>How it works:</p>
<ul>
<li><p>Input: You start with a list of genes of interest, often derived
from experimental data like differentially expressed genes (DEGs) from a
RNA-Seq experiment.</p></li>
<li><p>Annotation: Each gene in your list is associated with GO terms
that describe its known functions.</p></li>
<li><p>Statistical Test: A statistical test (e.g., Fisher’s exact test,
hypergeometric test) is applied to determine if any GO terms are
enriched in your gene list compared to what would be expected by
chance.</p></li>
<li><p>Output: The result is a list of GO terms along with their
associated p-values (or adjusted p-values). Significantly enriched terms
suggest that the biological processes they represent are particularly
relevant to your gene list.</p></li>
</ul>
<div id="categories-of-go-analysis" class="section level2">
<h2>Categories of GO Analysis</h2>
<p>The Gene Ontology is organized into three main categories:</p>
<ul>
<li><p>Biological Process (BP): Represents a series of events
accomplished by one or more ordered assemblies of molecular functions
(e.g., DNA replication, cell cycle, signal transduction).</p></li>
<li><p>Molecular Function (MF): Describes the elemental activities of a
gene product at the molecular level, such as binding or catalysis (e.g.,
kinase activity, DNA binding, transporter activity).</p></li>
<li><p>Cellular Component (CC): Refers to the location relative to
cellular structures in which a gene product performs a function (e.g.,
nucleus, cytoplasm, plasma membrane). Types of GO Enrichment
Analysis:</p></li>
</ul>
<p>For more information, we encourage further reading at: <a
href="https://geneontology.org/docs/ontology-documentation/"
class="uri">https://geneontology.org/docs/ontology-documentation/</a></p>
<p><strong>There are two main approaches to GO enrichment
analysis:</strong></p>
<ul>
<li><p>Over-Representation Analysis (ORA): This is the simpler method
and focuses on identifying GO terms that are significantly
over-represented in your gene list. It is based on the assumption that
all genes in the list are independent.</p></li>
<li><p>Functional Class Scoring (FCS): This approach considers the
relative ranking of genes in your list (e.g., based on expression
levels) and tests if gene sets associated with specific GO terms are
enriched towards the top or bottom of the list. This can reveal subtle
patterns that ORA might miss.</p></li>
</ul>
</div>
<div id="interpreting-results" class="section level2">
<h2>Interpreting Results:</h2>
<p>The enriched GO terms provide valuable insights into the biological
functions that are most relevant to your gene list. This can help
generate hypotheses about the underlying mechanisms involved and guide
further research.</p>
<div id="kyoto-encyclopedia-of-genes-and-genomes-kegg"
class="section level3">
<h3>Kyoto Encyclopedia of Genes and Genomes (KEGG)</h3>
<p>KEGG is a collection of manually drawn pathway maps representing
molecular interaction and reaction networks.</p>
</div>
<div id="other-gene-sets" class="section level3">
<h3>Other gene sets</h3>
<p>GO and KEGG are the most frequently used for functional analysis.
They are typically the first choice because of their long-standing
curation and availability for a wide range of species.</p>
<p>Other gene sets include but are not limited to Disease Ontology
<code>DO</code>, Disease Gene Network <code>DisGeNET</code>,
<code>wikiPathways</code>, Molecular Signatures Database
<code>MSigDb</code>.</p>
</div>
</div>
</div>
<div id="input-deg-list-and-deg-filtering" class="section level1">
<h1>Input DEG list and DEG filtering</h1>
<pre class="r"><code>Transplant_vs_Naive_annotated &lt;- read.csv(file = &quot;./results/Transplant_vs_Naive/Transplant_vs_Naive_annotated_DEGlist.csv&quot;,
                                          row.names = 1)

normalized_counts_annotated &lt;- read.csv(file=&quot;./results/Transplant_vs_Naive/normalized_counts.csv&quot;)</code></pre>
<p>For our enrichment analysis, we will use DEGs filtered by the
following cutoff:
<code>adj.P.Val &lt;= 0.05, logFC &lt;= -1 | logFC &gt;= 1</code>.</p>
<pre class="r"><code>Transplant_vs_Naive_annotated_DEGs &lt;- Transplant_vs_Naive_annotated %&gt;%
  dplyr::filter(str_trim(external_gene_name) != &quot;&quot;) %&gt;%  # Remove empty strings
  dplyr::filter(padj &lt;= 0.05, log2FoldChange &lt;= -1 | log2FoldChange &gt;= 1) %&gt;% # THIS IS FILTERED ANALYSIS
  dplyr::arrange(dplyr::desc(abs(log2FoldChange))) %&gt;%
  # Filter out the duplicated rows using `dplyr::distinct()`
  dplyr::distinct(external_gene_name, .keep_all = TRUE)</code></pre>
<p>Quick look at the number of DEGs.</p>
<pre class="r"><code>nrow(Transplant_vs_Naive_annotated_DEGs)</code></pre>
<pre><code>## [1] 1179</code></pre>
<p>Check for duplicated gene_ids</p>
<pre class="r"><code>any(duplicated(Transplant_vs_Naive_annotated_DEGs$external_gene_name))</code></pre>
<pre><code>## [1] FALSE</code></pre>
</div>
<div id="go-by-gprofiler2" class="section level1">
<h1>GO by <code>gprofiler2</code></h1>
<p><code>gprofiler2</code> performs functional enrichment analysis, also
known as over-representation analysis (ORA) on input gene list. It maps
genes to known functional information sources and detects statistically
significantly enriched terms.</p>
<p>In other words, <code>gprofiler2</code> takes a list of genes and
analyzes them to see if they are enriched for certain functions. This
can be helpful for understanding the biological processes that the genes
are involved in.</p>
<p>Many GO enrichment analysis tools allow you to archive your results
for future reference or sharing. <code>gprofiler2</code> maintains
archived versions based on genome versions used for secondary analysis,
specifically Ensembl. Thus, here we also set the archive to Ensembl 109
version.</p>
<p>Archiving results is often optional, but it can be a valuable
practice for reproducibility and collaboration.</p>
<pre class="r"><code>set_base_url(&quot;https://biit.cs.ut.ee/gprofiler_archive3/e109_eg56_p17&quot;)</code></pre>
<p>Further, note that although <code>gprofiler2</code> accepts Ensembl
IDs, most biologists/scientist prefer gene names as they are more
interpretable. One caveat, is that the use of gene names can can lead to
redundancies for cases where a name corresponds to more than one id
(e.g: haplotypes). However, for the most part it is acceptable to use
gene names. The example below is executed with gene names.</p>
<p>Lastly, note we set the <code>custom_bg</code> (custom background) to
the gene names present in our normalized counts which is derived from
the raw data that was pre-filtered. This is important due to the
following reasons:</p>
<ul>
<li><p>Statistical Significance: The background set defines the domain
of genes from which your input list was derived. By comparing the
frequency of GO terms in your gene list to their frequency in the
background set, <code>gprofiler2</code> can calculate p-values that
indicate the likelihood of observing such enrichment by chance. Without
a proper background set, these p-values would be unreliable.</p></li>
<li><p>Context-Specific Analysis: The choice of background set can
dramatically influence the results of your GO analysis. Using the entire
genome as the background might not be appropriate if your gene list
comes from a specific tissue, cell type, or experimental condition. In
such cases, it’s often more meaningful to use a custom background set
that reflects the relevant biological context (in our case this
corresponds to the total gene number after pre-filtering).</p></li>
<li><p>Controlling False Positives: A well-defined background set helps
control the false positive rate in your analysis. If your background set
is too small or biased, you might detect spurious enrichment of GO terms
that are not truly relevant to your gene list.</p></li>
</ul>
<p><code>gprofiler2</code> offers several options for choosing a
background set: annotated genes, custom background, and predefined
backgrounds.</p>
<pre class="r"><code># select species
species &lt;- &quot;mmusculus&quot; 

#select data sources, these are our min. standards:

data_sources &lt;- c(&quot;GO&quot;, &quot;KEGG&quot;, &quot;REAC&quot;, &quot;MIRNA&quot;, &quot;HP&quot;, &quot;HPA&quot;, &quot;WP&quot;)

gost_results &lt;- gost(query = Transplant_vs_Naive_annotated_DEGs$external_gene_name, 
                     organism = species,
                     ordered_query = FALSE, 
                     multi_query = FALSE,
                     significant = TRUE,
                     exclude_iea = FALSE, 
                     measure_underrepresentation = FALSE, 
                     user_threshold = 0.05,
                     correction_method = &quot;g_SCS&quot;, 
                     domain_scope = &quot;custom&quot;,
                     custom_bg = normalized_counts_annotated$external_gene_name, # genes names from normalized counts
                     numeric_ns = &quot;&quot;,
                     sources = data_sources,
                     as_short_link = FALSE,
                     evcodes = TRUE)</code></pre>
<p><strong>Challenge: Enable the url option for gost</strong></p>
<details>
<summary>
Click here for solution
</summary>
<p>explore the <code>gprofiler</code> results in detail through the
url.</p>
<pre class="r"><code>gost_results_url &lt;- gost(query = Transplant_vs_Naive_annotated_DEGs$external_gene_name, 
                         organism = species,
                         ordered_query = FALSE, 
                         multi_query = FALSE,
                         significant = TRUE,
                         exclude_iea = FALSE, 
                         measure_underrepresentation = FALSE, 
                         user_threshold = 0.05,
                         correction_method = &quot;g_SCS&quot;, 
                         domain_scope = &quot;custom&quot;,
                         custom_bg = normalized_counts_annotated$external_gene_name, # genes names from normalized counts
                         numeric_ns = &quot;&quot;,
                         sources = data_sources,
                         as_short_link = TRUE,
                         evcodes = TRUE)

gost_results_url</code></pre>
<pre><code>## [1] &quot;https://biit.cs.ut.ee/gplink/l/J4xuSIHGQY&quot;</code></pre>
</details>
</div>
<div id="explore-gprofiler2-results" class="section level1">
<h1>Explore <code>gprofiler2</code> results</h1>
<pre class="r"><code>gostplot(gost_results, capped = TRUE, interactive = TRUE)</code></pre>
<div class="plotly html-widget html-fill-item" id="htmlwidget-d74010a3a9c5b06c7f37" style="width:672px;height:480px;"></div>
<script type="application/json" data-for="htmlwidget-d74010a3a9c5b06c7f37">{"x":{"data":[{"x":[57.427862086106913,69.190689036069514,50.149145527476904,57.452789197609071,115.82931465660637,89.761095383461111,50.154684885588487,105.20205611951984,86.944331783717303,89.813719285521216,55.807599838466672,55.804830159410876,107.30701220192427,55.802060480355088,90.639083644148215,61.87042729160256,49.996813179408157,50.608912250738911,88.957888457280475,84.047247491355435,55.915617321642685,57.430631765162708,58.031652120270287,90.633544286036624,82.695644112127326,55.854684382415186,89.858034150413943,57.292147812372939,82.701183470238931,115.82654497755057,102.54870358406797,113.45016034767821,50.107600341639966,107.30424252286846,88.714156700370495,96.488645809987872,113.4584693848456,57.505413099669184,103.19680848312406,84.044477812299633,88.725235416593662,83.113865649552423,115.3418511427864,55.82698759185724,64.329902293148777,90.702786262431516,57.23952391031284,62.778882021903414,57.433401444218504,84.152495295475646,57.51372213683657,89.82202832268861,72.170863700105244,91.021299353847965,57.361389788767831,122.36852690733905,50.002352537519748,84.158034653587251,96.145205607069229,90.705555941487304,113.72435857420193,73.275965643367556,125.69491145334916,73.281505001479132,113.45293002673399,57.574655076064062,57.516491815892373,98.153222922520825,98.155992601576614,91.026838711959556,122.37129658639483,89.664156616508265,82.820279669638126,71.049143682508131,50.027279649021906,84.343603150325521,73.278735322423344,57.322614281986695,50.517512841897677,99.490977906469951,49.935880240180659,89.317946734533862,69.223925184739045,57.353080751600444,50.514743162841874,71.982525524311157,62.820427207740337,113.4612390639014,87.426255939425687,98.108908057628113,60.698853051001144,58.854246799841491,89.827567680800186,82.728880260796885,69.409493681477329,113.72158889514614,55.840835987136209,79.147685241653562,57.510952457780775,57.527570532115547,61.510369014349166,82.72334090268528,115.08704066965323,73.270426285255965,55.893459889196329,72.187481774440002,82.91998811564676,73.273195964311768,58.779465465335015,52.143314447649509,59.012118506021814,126.73077142021661,51.185005494344338,107.95788678003616,57.630048657179969,70.33733616516875,74.090251285771373,87.437334655648868,77.015032368691195,89.838646397023368,85.922320212128838,73.575090981393444,69.320863951691877,99.587916673422797,83.304973504402298,89.819258643632821,107.80555443196742,57.58573379228725,89.694623086122022,57.660515126793726,52.486754650568116,88.489812696851075,58.857016478897279,113.42523323617604,89.97990002886894,55.821448233745642,70.71401251675691,88.259929335220065,115.61881904836592,69.478735657872221,67.714450099330605,96.737916925009429,69.625528647829356,72.173633379161032,83.127714044831393,103.13587554389655,88.697538626035708,109.35657470321279,109.35103534510118,57.926404316150069,57.618969940956781,64.518240468942864,91.024069032903753,89.816488964577019,57.494334383446002,115.9234837445034,75.206431945256867,55.829757270913035,89.733398592903157,75.220280340535851,73.267656606200163,53.306579651083531,50.520282520953465,89.675235332731447,52.669553468250612,89.461970045435223,69.420572397700511,98.781940068186358,64.382526195208897,89.697392765177824,52.686171542585377,87.429025618481475,71.040834645340752,64.515470789887061,57.250602626536015,52.65293539391584,49.952498314515438,52.700019937864361,67.506724170145972,88.885876801829795,58.796083539669795,88.891416159941386,113.4556997057898,114.9208599263055,56.732672643102298,50.628300004129486,114.91809024724971,116.01211347428885,50.528591558120851,57.25337230559181,69.453808546370055,88.691999267924132,72.085003649375579,53.301040292971933,52.866200681212078,53.339815799753076,87.440104334704657,79.119988451095608,90.697246904319911,88.293165483889609,91.035147749126935,53.334276441641478,50.157454564644283,69.423342076756313,52.148853805761092,96.483106451876282,113.46400874295719,116.82916879574846,102.94476768904667,103.03616709788793,57.261681342759196,89.730628913847355,76.486023669034296,89.714010839512582,82.706722828350507,59.926112594434265,57.242293589368629,87.578588287494426,115.09258002776481,49.94695895640384],"y":[17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,17,15.726146737732263,15.594672000307513,15.56302551817681,15.16553655862409,14.492808438044868,14.187302943776974,14.173318472609385,14.121305986467053,13.252762206877515,13.027769085432766,12.804517229750676,12.486370314015968,12.41097694198149,12.331242215262554,12.323431110239156,12.138196297662422,12.049464151520201,12.037131739892454,11.96634496762389,11.887830960729115,11.886483446725487,11.736902863629929,11.621827446340189,11.430893583289004,11.380468764366723,11.203777202690302,11.028532630799987,10.944878142571241,10.906392910932594,10.906392910932594,10.906392910932594,10.906392910932594,10.687358374601921,10.669160147216401,10.639872209839275,10.639872209839275,10.639872209839275,10.619168176576709,10.619168176576709,10.617436839566043,10.61650437771377,10.341138866115175,9.649087289669076,9.6122449135025274,9.5600489113414024,9.4964681328108007,9.4059498779871795,9.3164862901529535,9.2360599200539202,9.1986142417973564,9.1437712467235084,9.0665865045163976,9.039223154881741,9.0246280847005096,8.8465183084060399,8.6296670802117479,8.5889286629941406,8.3927070633369389,8.3396258511191856,8.2934872111121507,8.1908556405422512,8.1467533480463743,8.0218440442184615,7.8651463037650204,7.7746839140102066,7.7319836080447173,7.7111929407153363,7.6666597435817003,7.6372594417103992,7.5599041513831242,7.5511261306716735,7.496686713653812,7.4590333686700223,7.4046345932475273,7.3870256471230391,7.3521162863284308,7.0742474848196997,6.9862743284713416,6.9804403716278101,6.9799161714394469,6.9572392580418585,6.8652443888584465,6.7988215683202293,6.7775571917953368,6.681005252083426,6.5228618305228618,6.4136514584564814,6.3325383975436473,6.2117509442348364,6.1366616073468512,6.0062015852830459,5.9944221167199121,5.9910408295367636,5.922757576228789,5.805452905490581,5.805452905490581,5.5231547136853436,5.5071126735916343,5.3650417561450174,5.2168244416402842,5.0732248482825044,5.0568115058883389,4.8111401487140428,4.8111401487140428,4.5012075288474778,4.1574378289412692,4.1374353231986793,4.1374353231986793,4.0561178723689491,4.0100519226225346,3.9004805528692619,3.8851846240148844,3.7957834170641283,3.7291159603915789,3.6632182028040337,3.6242035898573506,3.6242035898573506,3.594782978279194,3.5749332814404737,3.5661754148693348,3.5292243924482043,3.5007452686333225,3.4673779646746525,3.3379547453127252,3.3379547453127252,3.3343662100872575,3.3343662100872575,3.2595572866447835,3.2389128313511946,3.2148468967108963,3.1289071397503316,3.1289071397503316,3.112882148300157,3.1012537225138068,3.0681682863893807,3.0191659549972978,2.9409150160455089,2.933558868974619,2.8855880986951061,2.8299971109973483,2.8156745976737638,2.7805778017231915,2.7452371232741397,2.6702579150281767,2.6655992752639861,2.6641823632713955,2.6583742265757389,2.5692238588700733,2.5230925904230213,2.512247636989172,2.4932546286108668,2.4738144842109899,2.4738144842109899,2.4191162866060525,2.4004491822406009,2.3529487405178231,2.3236341681628008,2.3236341681628008,2.2961038104896621,2.2926429525709175,2.2716343659778673,2.2698951189845675,2.2697020159637411,2.258815966629796,2.2044238615404201,2.1822522653480476,2.1350187705896722,2.1305639544617709,2.1067395015298587,2.0959746715489334,2.0170939694319432,1.9686761372421029,1.9213858762753138,1.8505002067066436,1.8502858675625258,1.8129706666529439,1.8129706666529439,1.8129706666529439,1.7855090240310334,1.6845322763911481,1.6264556656865903,1.5304719500470667,1.4919875419166475,1.4788532873196405,1.4101100097492421,1.4070225947264141,1.3703256768005343,1.3129727608922546],"text":["GO:0007049 (1406) <br> cell cycle <br> 1.685e-55","GO:0022402 (1015) <br> cell cycle process <br> 3.586e-54","GO:0000278 (745) <br> mitotic cell cycle <br> 1.722e-48","GO:0007059 (352) <br> chromosome segregation <br> 1.219e-46","GO:1903047 (634) <br> mitotic cell cycle process <br> 1.363e-46","GO:0051276 (505) <br> chromosome organization <br> 7.949e-43","GO:0000280 (328) <br> nuclear division <br> 2.775e-39","GO:0098813 (265) <br> nuclear chromosome segregation <br> 2.779e-36","GO:0048285 (375) <br> organelle fission <br> 9.326e-34","GO:0051301 (507) <br> cell division <br> 4.236e-32","GO:0006261 (137) <br> DNA-templated DNA replication <br> 8.994e-32","GO:0006260 (253) <br> DNA replication <br> 2.542e-31","GO:0140014 (232) <br> mitotic nuclear division <br> 3.222e-31","GO:0006259 (871) <br> DNA metabolic process <br> 2.307e-30","GO:0051726 (872) <br> regulation of cell cycle <br> 1.071e-28","GO:0010564 (594) <br> regulation of cell cycle process <br> 6.058e-27","GO:0000070 (169) <br> mitotic sister chromatid segregation <br> 6.288e-27","GO:0000819 (204) <br> sister chromatid segregation <br> 3.029e-26","GO:0050896 (5479) <br> response to stimulus <br> 1.887e-23","GO:0045787 (274) <br> positive regulation of cell cycle <br> 4.532e-20","GO:0006310 (269) <br> DNA recombination <br> 6.374e-20","GO:0007051 (172) <br> spindle organization <br> 2.027e-19","GO:0007346 (403) <br> regulation of mitotic cell cycle <br> 2.350e-19","GO:0051716 (4621) <br> cellular response to stimulus <br> 8.997e-19","GO:0044770 (470) <br> cell cycle phase transition <br> 1.129e-18","GO:0006281 (519) <br> DNA repair <br> 2.033e-18","GO:0051321 (176) <br> meiotic cell cycle <br> 4.992e-18","GO:0006974 (763) <br> DNA damage response <br> 7.727e-18","GO:0044772 (386) <br> mitotic cell cycle phase transition <br> 8.832e-18","GO:1903046 (132) <br> meiotic cell cycle process <br> 2.923e-17","GO:0090068 (206) <br> positive regulation of cell cycle process <br> 3.636e-17","GO:1901987 (369) <br> regulation of cell cycle phase transition <br> 8.262e-17","GO:0000226 (501) <br> microtubule cytoskeleton organization <br> 1.879e-16","GO:0140013 (122) <br> meiotic nuclear division <br> 2.543e-16","GO:0050789 (7424) <br> regulation of biological process <br> 2.735e-16","GO:0065007 (7659) <br> biological regulation <br> 6.831e-16","GO:1901990 (289) <br> regulation of mitotic cell cycle phase transition <br> 3.215e-15","GO:0007088 (90) <br> regulation of mitotic nuclear division <br> 6.497e-15","GO:0090329 (48) <br> regulation of DNA-templated DNA replication <br> 6.709e-15","GO:0045786 (307) <br> negative regulation of cell cycle <br> 7.563e-15","GO:0050794 (6968) <br> regulation of cellular process <br> 5.588e-14","GO:0045132 (68) <br> meiotic chromosome segregation <br> 9.381e-14","GO:1902850 (140) <br> microtubule cytoskeleton organization involved in mitosis <br> 1.568e-13","GO:0006270 (29) <br> DNA replication initiation <br> 3.263e-13","GO:0016043 (4585) <br> cellular component organization <br> 3.882e-13","GO:0051783 (105) <br> regulation of nuclear division <br> 4.664e-13","GO:0006950 (2821) <br> response to stress <br> 4.749e-13","GO:0010948 (253) <br> negative regulation of cell cycle process <br> 7.275e-13","GO:0007052 (112) <br> mitotic spindle organization <br> 8.924e-13","GO:0045839 (51) <br> negative regulation of mitotic nuclear division <br> 9.181e-13","GO:0007093 (129) <br> mitotic cell cycle checkpoint signaling <br> 1.081e-12","GO:0051304 (73) <br> chromosome separation <br> 1.295e-12","GO:0032501 (4323) <br> multicellular organismal process <br> 1.299e-12","GO:0051983 (114) <br> regulation of chromosome segregation <br> 1.833e-12","GO:0007017 (688) <br> microtubule-based process <br> 2.389e-12","GO:1905818 (66) <br> regulation of chromosome separation <br> 3.708e-12","GO:0000075 (169) <br> cell cycle checkpoint signaling <br> 4.164e-12","GO:0045841 (46) <br> negative regulation of mitotic metaphase/anaphase transition <br> 6.255e-12","GO:0061982 (82) <br> meiosis I cell cycle process <br> 9.364e-12","GO:0051784 (55) <br> negative regulation of nuclear division <br> 1.135e-11","GO:1902100 (47) <br> negative regulation of metaphase/anaphase transition of cell cycle <br> 1.241e-11","GO:0033046 (47) <br> negative regulation of sister chromatid segregation <br> 1.241e-11","GO:2000816 (47) <br> negative regulation of mitotic sister chromatid separation <br> 1.241e-11","GO:0033048 (47) <br> negative regulation of mitotic sister chromatid segregation <br> 1.241e-11","GO:1901988 (221) <br> negative regulation of cell cycle phase transition <br> 2.054e-11","GO:0007127 (79) <br> meiosis I <br> 2.142e-11","GO:0007094 (44) <br> mitotic spindle assembly checkpoint signaling <br> 2.292e-11","GO:0071173 (44) <br> spindle assembly checkpoint signaling <br> 2.292e-11","GO:0071174 (44) <br> mitotic spindle checkpoint signaling <br> 2.292e-11","GO:0051985 (48) <br> negative regulation of chromosome segregation <br> 2.403e-11","GO:1905819 (48) <br> negative regulation of chromosome separation <br> 2.403e-11","GO:0051225 (116) <br> spindle assembly <br> 2.413e-11","GO:0044839 (139) <br> cell cycle G2/M phase transition <br> 2.418e-11","GO:0031577 (45) <br> spindle checkpoint signaling <br> 4.559e-11","GO:0000086 (123) <br> G2/M transition of mitotic cell cycle <br> 2.243e-10","GO:0045930 (192) <br> negative regulation of mitotic cell cycle <br> 2.442e-10","GO:0033047 (52) <br> regulation of mitotic sister chromatid segregation <br> 2.754e-10","GO:0006996 (2720) <br> organelle organization <br> 3.188e-10","GO:0000725 (149) <br> recombinational repair <br> 3.927e-10","GO:0071840 (4776) <br> cellular component organization or biogenesis <br> 4.825e-10","GO:0000003 (860) <br> reproduction <br> 5.807e-10","GO:0051052 (462) <br> regulation of DNA metabolic process <br> 6.330e-10","GO:0022414 (853) <br> reproductive process <br> 7.182e-10","GO:0007010 (1065) <br> cytoskeleton organization <br> 8.579e-10","GO:0000724 (146) <br> double-strand break repair via homologous recombination <br> 9.136e-10","GO:0032392 (78) <br> DNA geometric change <br> 9.449e-10","GO:0010965 (55) <br> regulation of mitotic sister chromatid separation <br> 1.424e-09","GO:1901991 (156) <br> negative regulation of mitotic cell cycle phase transition <br> 2.346e-09","GO:0048518 (4416) <br> positive regulation of biological process <br> 2.577e-09","GO:0071103 (87) <br> DNA conformation change <br> 4.048e-09","GO:0009987 (10552) <br> cellular process <br> 4.575e-09","GO:0008283 (1338) <br> cell population proliferation <br> 5.088e-09","GO:0051306 (58) <br> mitotic sister chromatid separation <br> 6.444e-09","GO:0044786 (33) <br> cell cycle DNA replication <br> 7.133e-09","GO:0030071 (84) <br> regulation of mitotic metaphase/anaphase transition <br> 9.509e-09","GO:1902099 (85) <br> regulation of metaphase/anaphase transition of cell cycle <br> 1.364e-08","GO:0006275 (120) <br> regulation of DNA replication <br> 1.680e-08","GO:0042127 (1105) <br> regulation of cell population proliferation <br> 1.854e-08","GO:0007091 (86) <br> metaphase/anaphase transition of mitotic cell cycle <br> 1.944e-08","GO:0007098 (127) <br> centrosome cycle <br> 2.154e-08","GO:0010389 (92) <br> regulation of G2/M transition of mitotic cell cycle <br> 2.305e-08","GO:0044784 (87) <br> metaphase/anaphase transition of cell cycle <br> 2.755e-08","GO:1902749 (104) <br> regulation of cell cycle G2/M phase transition <br> 2.811e-08","GO:0033044 (230) <br> regulation of chromosome organization <br> 3.186e-08","GO:0006302 (260) <br> double-strand break repair <br> 3.475e-08","GO:0032508 (72) <br> DNA duplex unwinding <br> 3.939e-08","GO:0045005 (48) <br> DNA-templated DNA replication maintenance of fidelity <br> 4.102e-08","GO:0033045 (94) <br> regulation of sister chromatid segregation <br> 4.445e-08","GO:0008150 (12289) <br> biological_process <br> 8.429e-08","GO:0002250 (337) <br> adaptive immune response <br> 1.032e-07","GO:0008608 (41) <br> attachment of spindle microtubules to kinetochore <br> 1.046e-07","GO:2001251 (91) <br> negative regulation of chromosome organization <br> 1.047e-07","GO:0001775 (826) <br> cell activation <br> 1.103e-07","GO:0140694 (323) <br> non-membrane-bounded organelle assembly <br> 1.364e-07","GO:0007154 (3799) <br> cell communication <br> 1.589e-07","GO:0031023 (135) <br> microtubule organizing center organization <br> 1.669e-07","GO:0033554 (1605) <br> cellular response to stress <br> 2.084e-07","GO:0048522 (3947) <br> positive regulation of cellular process <br> 3.000e-07","GO:0035825 (39) <br> homologous recombination <br> 3.858e-07","GO:0051310 (68) <br> metaphase plate congression <br> 4.650e-07","GO:0046649 (630) <br> lymphocyte activation <br> 6.141e-07","GO:0033260 (28) <br> nuclear DNA replication <br> 7.300e-07","GO:0023052 (3727) <br> signaling <br> 9.858e-07","GO:0071897 (163) <br> DNA biosynthetic process <br> 1.013e-06","GO:0045321 (748) <br> leukocyte activation <br> 1.021e-06","GO:0051303 (76) <br> establishment of chromosome localization <br> 1.195e-06","GO:0140527 (33) <br> reciprocal homologous recombination <br> 1.565e-06","GO:0007131 (33) <br> reciprocal meiotic recombination <br> 1.565e-06","GO:0051239 (2002) <br> regulation of multicellular organismal process <br> 2.998e-06","GO:0007165 (3460) <br> signal transduction <br> 3.111e-06","GO:0002376 (1703) <br> immune system process <br> 4.315e-06","GO:0050000 (81) <br> chromosome localization <br> 6.070e-06","GO:0008284 (622) <br> positive regulation of cell population proliferation <br> 8.448e-06","GO:1901976 (45) <br> regulation of cell cycle checkpoint <br> 8.774e-06","GO:0051383 (21) <br> kinetochore organization <br> 1.545e-05","GO:0006268 (21) <br> DNA unwinding involved in DNA replication <br> 1.545e-05","GO:0031297 (43) <br> replication fork processing <br> 3.153e-05","GO:0048856 (3677) <br> anatomical structure development <br> 6.959e-05","GO:1902969 (13) <br> mitotic DNA replication <br> 7.287e-05","GO:0030174 (13) <br> regulation of DNA-templated DNA replication initiation <br> 7.287e-05","GO:0019953 (593) <br> sexual reproduction <br> 8.788e-05","GO:0070192 (36) <br> chromosome organization involved in meiotic cell cycle <br> 9.771e-05","GO:0030261 (32) <br> chromosome condensation <br> 1.258e-04","GO:0032502 (4022) <br> developmental process <br> 1.303e-04","GO:0045143 (37) <br> homologous chromosome segregation <br> 1.600e-04","GO:0090307 (69) <br> mitotic spindle assembly <br> 1.866e-04","GO:0050778 (510) <br> positive regulation of immune response <br> 2.172e-04","GO:1900264 (14) <br> positive regulation of DNA-directed DNA polymerase activity <br> 2.376e-04","GO:1900262 (14) <br> regulation of DNA-directed DNA polymerase activity <br> 2.376e-04","GO:0007292 (101) <br> female gamete generation <br> 2.542e-04","GO:0007143 (29) <br> female meiotic nuclear division <br> 2.661e-04","GO:0016126 (43) <br> sterol biosynthetic process <br> 2.715e-04","GO:0051984 (25) <br> positive regulation of chromosome segregation <br> 2.956e-04","GO:0051302 (129) <br> regulation of cell division <br> 3.157e-04","GO:0007080 (54) <br> mitotic metaphase plate congression <br> 3.409e-04","GO:1903083 (18) <br> protein localization to condensed chromosome <br> 4.592e-04","GO:0034501 (18) <br> protein localization to kinetochore <br> 4.592e-04","GO:0006271 (9) <br> DNA strand elongation involved in DNA replication <br> 4.631e-04","GO:0051256 (9) <br> mitotic spindle midzone assembly <br> 4.631e-04","GO:0034508 (26) <br> centromere complex assembly <br> 5.501e-04","GO:0033043 (976) <br> regulation of organelle organization <br> 5.769e-04","GO:0002684 (774) <br> positive regulation of immune system process <br> 6.098e-04","GO:0000727 (12) <br> double-strand break repair via break-induced replication <br> 7.432e-04","GO:0051231 (12) <br> spindle elongation <br> 7.432e-04","GO:0002449 (225) <br> lymphocyte mediated immunity <br> 7.711e-04","GO:0051128 (1890) <br> regulation of cellular component organization <br> 7.920e-04","GO:0030097 (759) <br> hemopoiesis <br> 8.547e-04","GO:0071459 (41) <br> protein localization to chromosome, centromeric region <br> 9.568e-04","GO:0016064 (101) <br> immunoglobulin mediated immune response <br> 1.146e-03","GO:0051240 (1166) <br> positive regulation of multicellular organismal process <br> 1.165e-03","GO:0002455 (32) <br> humoral immune response mediated by circulating immunoglobulin <br> 1.301e-03","GO:0048519 (3831) <br> negative regulation of biological process <br> 1.479e-03","GO:0031570 (116) <br> DNA integrity checkpoint signaling <br> 1.529e-03","GO:0016125 (96) <br> sterol metabolic process <br> 1.657e-03","GO:0006955 (1100) <br> immune response <br> 1.798e-03","GO:0002443 (300) <br> leukocyte mediated immunity <br> 2.137e-03","GO:0000022 (10) <br> mitotic spindle elongation <br> 2.160e-03","GO:0002460 (241) <br> adaptive immune response based on somatic recombination of immune receptors built from immunoglobulin superfamily domains <br> 2.167e-03","GO:0019724 (104) <br> B cell mediated immunity <br> 2.196e-03","GO:0050865 (502) <br> regulation of cell activation <br> 2.696e-03","GO:0008203 (92) <br> cholesterol metabolic process <br> 2.999e-03","GO:0050867 (312) <br> positive regulation of cell activation <br> 3.074e-03","GO:1901989 (99) <br> positive regulation of cell cycle phase transition <br> 3.212e-03","GO:1902653 (39) <br> secondary alcohol biosynthetic process <br> 3.359e-03","GO:0006695 (39) <br> cholesterol biosynthetic process <br> 3.359e-03","GO:0000910 (150) <br> cytokinesis <br> 3.810e-03","GO:1902652 (100) <br> secondary alcohol metabolic process <br> 3.977e-03","GO:1903131 (387) <br> mononuclear cell differentiation <br> 4.437e-03","GO:0000731 (30) <br> DNA synthesis involved in DNA repair <br> 4.746e-03","GO:0006956 (30) <br> complement activation <br> 4.746e-03","GO:0030154 (2709) <br> cell differentiation <br> 5.057e-03","GO:0050776 (636) <br> regulation of immune response <br> 5.097e-03","GO:0032465 (75) <br> regulation of cytokinesis <br> 5.350e-03","GO:0002682 (1078) <br> regulation of immune system process <br> 5.372e-03","GO:0002521 (501) <br> leukocyte differentiation <br> 5.374e-03","GO:0002696 (300) <br> positive regulation of leukocyte activation <br> 5.510e-03","GO:0048523 (3442) <br> negative regulation of cellular process <br> 6.246e-03","GO:0042113 (225) <br> B cell activation <br> 6.573e-03","GO:0051781 (52) <br> positive regulation of cell division <br> 7.328e-03","GO:0048869 (2732) <br> cellular developmental process <br> 7.403e-03","GO:0051988 (18) <br> regulation of attachment of spindle microtubules to kinetochore <br> 7.821e-03","GO:0002694 (468) <br> regulation of leukocyte activation <br> 8.017e-03","GO:0000281 (71) <br> mitotic cytokinesis <br> 9.614e-03","GO:0030098 (342) <br> lymphocyte differentiation <br> 1.075e-02","GO:0002252 (454) <br> immune effector process <br> 1.198e-02","GO:0065004 (174) <br> protein-DNA complex assembly <br> 1.411e-02","GO:1901992 (79) <br> positive regulation of mitotic cell cycle phase transition <br> 1.412e-02","GO:1903504 (19) <br> regulation of mitotic spindle checkpoint <br> 1.538e-02","GO:0090231 (19) <br> regulation of spindle checkpoint <br> 1.538e-02","GO:0090266 (19) <br> regulation of mitotic cell cycle spindle assembly checkpoint <br> 1.538e-02","GO:0006959 (100) <br> humoral immune response <br> 1.639e-02","GO:0051255 (12) <br> spindle midzone assembly <br> 2.068e-02","GO:0035556 (1872) <br> intracellular signal transduction <br> 2.363e-02","GO:0051249 (399) <br> regulation of lymphocyte activation <br> 2.948e-02","GO:0044774 (76) <br> mitotic DNA integrity checkpoint signaling <br> 3.221e-02","GO:0009605 (1824) <br> response to external stimulus <br> 3.320e-02","GO:0006952 (1117) <br> defense response <br> 3.889e-02","GO:0048583 (2724) <br> regulation of response to stimulus <br> 3.917e-02","GO:1902751 (30) <br> positive regulation of cell cycle G2/M phase transition <br> 4.263e-02","GO:0000018 (113) <br> regulation of DNA recombination <br> 4.864e-02"],"key":["GO:0007049","GO:0022402","GO:0000278","GO:0007059","GO:1903047","GO:0051276","GO:0000280","GO:0098813","GO:0048285","GO:0051301","GO:0006261","GO:0006260","GO:0140014","GO:0006259","GO:0051726","GO:0010564","GO:0000070","GO:0000819","GO:0050896","GO:0045787","GO:0006310","GO:0007051","GO:0007346","GO:0051716","GO:0044770","GO:0006281","GO:0051321","GO:0006974","GO:0044772","GO:1903046","GO:0090068","GO:1901987","GO:0000226","GO:0140013","GO:0050789","GO:0065007","GO:1901990","GO:0007088","GO:0090329","GO:0045786","GO:0050794","GO:0045132","GO:1902850","GO:0006270","GO:0016043","GO:0051783","GO:0006950","GO:0010948","GO:0007052","GO:0045839","GO:0007093","GO:0051304","GO:0032501","GO:0051983","GO:0007017","GO:1905818","GO:0000075","GO:0045841","GO:0061982","GO:0051784","GO:1902100","GO:0033046","GO:2000816","GO:0033048","GO:1901988","GO:0007127","GO:0007094","GO:0071173","GO:0071174","GO:0051985","GO:1905819","GO:0051225","GO:0044839","GO:0031577","GO:0000086","GO:0045930","GO:0033047","GO:0006996","GO:0000725","GO:0071840","GO:0000003","GO:0051052","GO:0022414","GO:0007010","GO:0000724","GO:0032392","GO:0010965","GO:1901991","GO:0048518","GO:0071103","GO:0009987","GO:0008283","GO:0051306","GO:0044786","GO:0030071","GO:1902099","GO:0006275","GO:0042127","GO:0007091","GO:0007098","GO:0010389","GO:0044784","GO:1902749","GO:0033044","GO:0006302","GO:0032508","GO:0045005","GO:0033045","GO:0008150","GO:0002250","GO:0008608","GO:2001251","GO:0001775","GO:0140694","GO:0007154","GO:0031023","GO:0033554","GO:0048522","GO:0035825","GO:0051310","GO:0046649","GO:0033260","GO:0023052","GO:0071897","GO:0045321","GO:0051303","GO:0140527","GO:0007131","GO:0051239","GO:0007165","GO:0002376","GO:0050000","GO:0008284","GO:1901976","GO:0051383","GO:0006268","GO:0031297","GO:0048856","GO:1902969","GO:0030174","GO:0019953","GO:0070192","GO:0030261","GO:0032502","GO:0045143","GO:0090307","GO:0050778","GO:1900264","GO:1900262","GO:0007292","GO:0007143","GO:0016126","GO:0051984","GO:0051302","GO:0007080","GO:1903083","GO:0034501","GO:0006271","GO:0051256","GO:0034508","GO:0033043","GO:0002684","GO:0000727","GO:0051231","GO:0002449","GO:0051128","GO:0030097","GO:0071459","GO:0016064","GO:0051240","GO:0002455","GO:0048519","GO:0031570","GO:0016125","GO:0006955","GO:0002443","GO:0000022","GO:0002460","GO:0019724","GO:0050865","GO:0008203","GO:0050867","GO:1901989","GO:1902653","GO:0006695","GO:0000910","GO:1902652","GO:1903131","GO:0000731","GO:0006956","GO:0030154","GO:0050776","GO:0032465","GO:0002682","GO:0002521","GO:0002696","GO:0048523","GO:0042113","GO:0051781","GO:0048869","GO:0051988","GO:0002694","GO:0000281","GO:0030098","GO:0002252","GO:0065004","GO:1901992","GO:1903504","GO:0090231","GO:0090266","GO:0006959","GO:0051255","GO:0035556","GO:0051249","GO:0044774","GO:0009605","GO:0006952","GO:0048583","GO:1902751","GO:0000018"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(255,153,0,1)","opacity":0.80000000000000004,"size":[19.922344540698148,19.46726810900396,19.022814703642123,17.887302757651266,18.785727241444278,18.444911311714872,17.775605068912441,17.432701703048981,17.986671626313715,18.450900462417568,16.312747481161178,17.357053663760201,17.214455995464622,19.248977085488537,19.250625634851332,18.688889327254223,16.679643677666306,16.999963389973509,21.69750233766046,17.486963681195082,17.457068763981997,16.709924825529853,18.098885053072063,21.484979551523733,18.335647887786703,18.486296663591244,16.749381745878246,19.057585561962338,18.031830085801257,16.246640206736487,17.016355390127128,17.961415634119447,18.43285410440377,16.105389877305754,22.070462997812779,22.10829225242944,17.573115427111336,15.544045881826902,14.289551053282345,17.670133746427048,21.993271790565693,15.001993266891569,16.351103106737206,13.163593405554227,21.475158805409013,15.831759381153528,20.854210279116561,17.357053663760201,15.950197129898299,14.416985747273221,16.205587015350208,15.141645611173489,21.401098182181613,15.982477622673839,18.906311100912252,14.94271233506351,16.679643677666306,14.199158443904556,15.366838671154913,14.573601151658657,14.244933179291202,14.244933179291202,14.244933179291202,14.244933179291202,17.13385386455327,15.295124283351443,14.103902153872516,14.103902153872516,14.103902153872516,14.289551053282345,14.289551053282345,16.01411372552489,16.338422958806614,14.152168878418365,16.120101385980103,16.897644828791304,14.457481546757879,20.806684328326707,16.460776794464021,21.526347604456319,19.230705322275302,18.309410053163983,19.218944325833306,19.535249066274631,16.425077371417782,15.270510325237122,14.573601151658657,16.540988618015533,21.42792442408879,15.479821521486015,22.492831470812956,19.853945066281881,14.682412279994908,13.464799414642561,15.412969781594908,15.435558208980297,16.075547696656944,19.587176083598308,15.457839510796521,16.17760674094227,15.585496862893729,15.479821521486015,15.814098793219157,17.200124735719086,17.401657427958508,15.114631326484991,14.289551053282345,15.62591596323723,22.672961304097353,17.818526375131473,13.950740632328149,15.564903474122932,19.172570852896172,17.751189464361872,21.237361105077706,16.286640532720444,20.103588117735864,21.285950341600966,13.840863529293893,15.001993266891569,18.776349966356563,13.080107699525154,21.212983456951857,16.617210370299976,19.028673349481412,15.220157204576269,13.464799414642561,13.464799414642561,20.401801761469258,21.11795303985895,20.184086523542867,15.34327872942024,18.757397570249161,14.152168878418365,12.365130105867371,12.365130105867371,14.05429113859868,21.195755306834311,11.0179178849601,11.0179178849601,18.686377300767166,13.662461030030807,13.393923550484519,21.309832524076565,13.723889277645902,15.030869316627983,18.459835489562984,11.242006203757187,11.242006203757187,15.759918195883458,13.163593405554227,14.05429113859868,12.805215053656443,16.205587015350208,14.535751690269857,11.956334427574175,11.956334427574175,9.7835234899655621,9.7835234899655621,12.901287144618975,19.411657681609903,19.078394969223684,10.767816834397534,10.767816834397534,17.163675224929634,20.324643736102789,19.049936922258642,13.950740632328149,15.759918195883458,19.662548606866903,13.393923550484519,21.248036947720749,16.01411372552489,15.66535093296908,19.580798152828777,17.63317792996498,10.161652391358336,17.277273632652467,15.814098793219157,18.435878338912623,15.585496862893729,17.695959996502381,15.722754841180334,13.840863529293893,13.840863529293893,16.472495106982581,15.741443860532286,18.035864417001708,13.243549133997096,13.243549133997096,20.801393836186787,18.790391533312505,15.194396009723683,19.552355119326325,18.43285410440377,17.63317792996498,21.111265445269087,17.163675224929634,14.457481546757879,20.812429565981574,11.956334427574175,18.329134954281596,15.087173309405685,17.841823859838922,18.282665174496906,16.729781696869093,15.295124283351443,12.102003142235581,12.102003142235581,12.102003142235581,15.741443860532286,10.767816834397534,20.311783370303967,18.083393392381662,15.220157204576269,20.276824427386519,19.602355775854623,20.808602434543204,13.243549133997096,15.966419482055739],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(255,153,0,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"GO:BP","legendgroup":"GO:BP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[36.747336908689746,36.226527166172147,44.975022734632645,36.198824520293549,41.182530513852903,41.193611572204347,36.204365049469274,36.063081555488431,36.988349927833525,37.797267187488522,36.201594784881408,36.666999235641818,45.515224329265259,37.927469623117922,36.506323889545961,37.777875335373501,43.792119755616611,36.980039134069948,45.240968135067163,37.071457865469313,36.509094154133827,36.603283150121044,36.974498604894229,37.107471305111488,36.329026955922949,36.229297430760006,43.720092876332266,43.448606946722023,36.295783780868632,37.135173950990087,36.215446107820711,45.310224749763655,38.969089108153113,36.633756060587501,36.65868844187824,39.36800720880489,44.817117653124647,45.307454485175789,45.243738399655022,40.919355378006252,41.315503214070162,37.113011834287207,42.201987882185222,39.024494399910303,36.234837959935724,41.199152101380065,39.681047107233027,41.104963105392834,37.110241569699348],"y":[17,17,17,17,17,17,14.771958107531162,14.49101134768336,13.712250031034564,13.514009417101672,13.512506741680488,12.328833069966556,12.176770057299152,11.471918540324452,10.623881199607208,10.506800833588196,9.7977849097018463,8.1541459256167617,7.4155500132469605,6.9479885232132963,6.884461275853667,6.5854797452423384,6.5682862536228415,6.4747098476848972,6.1813248649216694,5.7507488696316686,5.5305694572518673,5.4906817432838491,5.4788400002871667,5.2564548966440956,4.8708389320916776,4.8489888142859812,4.7434029387141505,4.6931009811949007,4.341961673833306,4.2105198217585835,4.1669937850429015,4.0528058048667797,3.9126872230575924,3.7004148205199261,3.3924888839555187,2.8620419051690269,2.4782891541446923,2.2641732775148431,2.2641732775148431,2.0582492984275604,2.0561474742124961,1.9834507769746994,1.3129727608922546],"text":["GO:0005694 (1117) <br> chromosome <br> 1.939e-40","GO:0000793 (239) <br> condensed chromosome <br> 8.874e-27","GO:0098687 (329) <br> chromosomal region <br> 2.984e-25","GO:0000775 (232) <br> chromosome, centromeric region <br> 2.021e-21","GO:0043228 (3678) <br> non-membrane-bounded organelle <br> 1.276e-18","GO:0043232 (3678) <br> intracellular non-membrane-bounded organelle <br> 1.276e-18","GO:0000779 (160) <br> condensed chromosome, centromeric region <br> 1.691e-15","GO:0000228 (204) <br> nuclear chromosome <br> 3.228e-15","GO:0005819 (377) <br> spindle <br> 1.940e-14","GO:0009986 (568) <br> cell surface <br> 3.062e-14","GO:0000776 (152) <br> kinetochore <br> 3.073e-14","GO:0005657 (71) <br> replication fork <br> 4.690e-13","GO:0110165 (11503) <br> cellular anatomical entity <br> 6.656e-13","GO:0015630 (1114) <br> microtubule cytoskeleton <br> 3.374e-12","GO:0005575 (12333) <br> cellular_component <br> 2.377e-11","GO:0009897 (279) <br> external side of plasma membrane <br> 3.113e-11","GO:0072686 (158) <br> mitotic spindle <br> 1.593e-10","GO:0005815 (711) <br> microtubule organizing center <br> 7.012e-09","GO:0099080 (938) <br> supramolecular complex <br> 3.841e-08","GO:0005856 (1690) <br> cytoskeleton <br> 1.127e-07","GO:0005576 (876) <br> extracellular region <br> 1.305e-07","GO:0005615 (630) <br> extracellular space <br> 2.597e-07","GO:0005813 (567) <br> centrosome <br> 2.702e-07","GO:0005874 (362) <br> microtubule <br> 3.352e-07","GO:0000940 (15) <br> outer kinetochore <br> 6.587e-07","GO:0000794 (56) <br> condensed nuclear chromosome <br> 1.775e-06","GO:0071944 (3094) <br> cell periphery <br> 2.947e-06","GO:0071162 (11) <br> CMG complex <br> 3.231e-06","GO:0000922 (155) <br> spindle pole <br> 3.320e-06","GO:0005886 (2877) <br> plasma membrane <br> 5.540e-06","GO:0000785 (618) <br> chromatin <br> 1.346e-05","GO:0099513 (499) <br> polymeric cytoskeletal fiber <br> 1.416e-05","GO:0031261 (12) <br> DNA replication preinitiation complex <br> 1.805e-05","GO:0005634 (5638) <br> nucleus <br> 2.027e-05","GO:0005654 (3071) <br> nucleoplasm <br> 4.550e-05","GO:0031981 (3470) <br> nuclear lumen <br> 6.159e-05","GO:0098552 (452) <br> side of membrane <br> 6.808e-05","GO:0099512 (631) <br> supramolecular fiber <br> 8.855e-05","GO:0099081 (635) <br> supramolecular polymer <br> 1.223e-04","GO:0042555 (11) <br> MCM complex <br> 1.993e-04","GO:0043596 (39) <br> nuclear replication fork <br> 4.051e-04","GO:0005876 (76) <br> spindle microtubule <br> 1.374e-03","GO:0051233 (34) <br> spindle midzone <br> 3.324e-03","GO:0031390 (8) <br> Ctf18 RFC-like complex <br> 5.443e-03","GO:0000796 (8) <br> condensin complex <br> 5.443e-03","GO:0043235 (219) <br> receptor complex <br> 8.745e-03","GO:0032991 (4370) <br> protein-containing complex <br> 8.787e-03","GO:0043073 (59) <br> germ cell nucleus <br> 1.039e-02","GO:0005875 (113) <br> microtubule associated complex <br> 4.864e-02"],"key":["GO:0005694","GO:0000793","GO:0098687","GO:0000775","GO:0043228","GO:0043232","GO:0000779","GO:0000228","GO:0005819","GO:0009986","GO:0000776","GO:0005657","GO:0110165","GO:0015630","GO:0005575","GO:0009897","GO:0072686","GO:0005815","GO:0099080","GO:0005856","GO:0005576","GO:0005615","GO:0005813","GO:0005874","GO:0000940","GO:0000794","GO:0071944","GO:0071162","GO:0000922","GO:0005886","GO:0000785","GO:0099513","GO:0031261","GO:0005634","GO:0005654","GO:0031981","GO:0098552","GO:0099512","GO:0099081","GO:0042555","GO:0043596","GO:0005876","GO:0051233","GO:0031390","GO:0000796","GO:0043235","GO:0032991","GO:0043073","GO:0005875"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(16,150,24,1)","opacity":0.80000000000000004,"size":[19.602355775854623,17.263544337493027,17.780438448899918,17.214455995464622,21.196102328187639,21.196102328187639,16.585008068656709,16.999963389973509,17.994990730367135,18.622015985711712,16.495667307919501,15.087173309405685,22.595044252200768,19.598577558116688,22.677165354330711,17.516254668148232,16.5631565955727,18.954551180399363,19.35509052362724,20.173700917358246,19.257199234067556,18.776349966356563,18.619377021884127,17.931367500024635,11.444741251072838,14.610639892328045,20.974033235279048,10.485368589139895,16.529783183148883,20.879778421573118,18.74782064896047,18.426785601269309,10.767816834397534,21.732952489696334,20.96438548112959,21.121652244689187,18.275897504535966,18.778700405544626,18.788061404908159,10.485368589139895,13.840863529293893,15.220157204576269,13.533067077450161,9.3304016728025392,9.3304016728025392,17.118714701686411,21.414731988385057,14.717204984820066,15.966419482055739],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(16,150,24,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"GO:CC","legendgroup":"GO:CC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[6.1408888386478298,6.1076509349262418,10.774806582499213,6.2904594053949756,2.8891139245524773,31.288132829339229,2.9306613042044622,2.8974234004828743,6.2267534232619317,2.1080231870951609,2.8946535751727418,2.9057328764132713,27.460234250736356,2.9943672863375057,9.3871241021229181,25.443801424960018,17.344832218133092,5.1797594560319125,17.037381608708408,31.703606625859084,10.727719552226963,2.9140423523436683,17.738147412171884,13.007285782465869,32.384983652151632,2.0830947593039699,16.912739469752452,6.1630474411288887,4.0662896813587164],"y":[15.65193719912928,13.088485303578658,8.0714074616586338,7.4057751240393621,7.1473999729118551,7.1253892221791304,6.2013753477011511,5.2058468965036928,3.8604418949688579,3.6688545331116051,3.5730306503279348,3.2372078966356383,3.1492268819951983,2.7216336723296011,2.6901406579846507,2.4459051515774322,2.4459051515774322,2.3085603602135398,2.052624953743571,1.973051511763896,1.7888139233650986,1.6406930455715154,1.63630898655669,1.5278708778907613,1.5121056581799768,1.5121056581799768,1.4975250089625813,1.4324972892593124,1.3944336295573487],"text":["GO:0005515 (6957) <br> protein binding <br> 2.229e-16","GO:0005488 (9575) <br> binding <br> 8.157e-14","GO:0017116 (23) <br> single-stranded DNA helicase activity <br> 8.484e-09","GO:0008094 (117) <br> ATP-dependent activity, acting on DNA <br> 3.928e-08","GO:0003674 (12222) <br> molecular_function <br> 7.122e-08","GO:0140097 (213) <br> catalytic activity, acting on DNA <br> 7.492e-08","GO:0003697 (115) <br> single-stranded DNA binding <br> 6.290e-07","GO:0003678 (54) <br> DNA helicase activity <br> 6.225e-06","GO:0008017 (213) <br> microtubule binding <br> 1.379e-04","GO:0000217 (33) <br> DNA secondary structure binding <br> 2.144e-04","GO:0003677 (1688) <br> DNA binding <br> 2.673e-04","GO:0003682 (484) <br> chromatin binding <br> 5.792e-04","GO:0097367 (1676) <br> carbohydrate derivative binding <br> 7.092e-04","GO:0003777 (48) <br> microtubule motor activity <br> 1.898e-03","GO:0015631 (291) <br> tubulin binding <br> 2.041e-03","GO:0060089 (496) <br> molecular transducer activity <br> 3.582e-03","GO:0038023 (496) <br> signaling receptor activity <br> 3.582e-03","GO:0004888 (379) <br> transmembrane signaling receptor activity <br> 4.914e-03","GO:0036094 (1949) <br> small molecule binding <br> 8.859e-03","GO:0140657 (462) <br> ATP-dependent activity <br> 1.064e-02","GO:0017076 (1592) <br> purine nucleotide binding <br> 1.626e-02","GO:0003689 (9) <br> DNA clamp loader activity <br> 2.287e-02","GO:0042802 (1679) <br> identical protein binding <br> 2.310e-02","GO:0030554 (1307) <br> adenyl nucleotide binding <br> 2.966e-02","GO:1901265 (1720) <br> nucleoside phosphate binding <br> 3.075e-02","GO:0000166 (1720) <br> nucleotide binding <br> 3.075e-02","GO:0035639 (1452) <br> purine ribonucleoside triphosphate binding <br> 3.180e-02","GO:0005524 (1170) <br> ATP binding <br> 3.694e-02","GO:0004386 (142) <br> helicase activity <br> 4.032e-02"],"key":["GO:0005515","GO:0005488","GO:0017116","GO:0008094","GO:0003674","GO:0140097","GO:0003697","GO:0003678","GO:0008017","GO:0000217","GO:0003677","GO:0003682","GO:0097367","GO:0003777","GO:0015631","GO:0060089","GO:0038023","GO:0004888","GO:0036094","GO:0140657","GO:0017076","GO:0003689","GO:0042802","GO:0030554","GO:1901265","GO:0000166","GO:0035639","GO:0005524","GO:0004386"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(220,57,18,1)","opacity":0.80000000000000004,"size":[21.991343739108068,22.377073975868932,12.597492423457624,16.029697700807695,22.666528872289991,17.072346651004239,15.998374687635199,14.535751690269857,17.072346651004239,13.464799414642561,20.172095457545684,18.380398134066112,20.162419240486116,14.289551053282345,17.584223761228237,18.417632333398242,18.417632333398242,18.003260990705325,20.365887149912243,18.309410053163983,20.092511183395217,9.7835234899655621,20.164845311038249,19.821499416155511,20.197538939770538,20.197538939770538,19.966609091687481,19.667340269756686,16.376156308132881],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(220,57,18,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"GO:MF","legendgroup":"GO:MF","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[158.43465581211814,175.52674313755094,193.80151095041847,150.69322264495159,160.2765352597624,156.38504559970195,180.91943379102975,169.12863558257609,193.53284582798014,155.98897227487018,183.52576245303462,164.92139516006233,151.69033031585678,155.29099690523657,160.04941629027846,172.64897960957742,154.52654769087593,161.70572292139315,168.65777918242642,155.40732613350883,167.44740125968877,173.90090368526944,149.96478009648479,167.45571049027964,155.51534613119023,150.00909599296943,176.23302773777542,160.00510039379378,154.52931743440621,165.04603361892549,175.16667647861297,172.91764473201573,169.10924737786405,153.81749334712114,163.36756903956845,161.72511112610522,163.81903723500608,150.25006368010486,161.71957163904463,155.0971148581161,153.09736002924518,150.95357853679906,153.72055232356092,170.46919145123749,160.05495577733905,170.44980324652545,153.60976258234922,155.68153074300776,151.04221032976841,193.10907506784542,153.69285488825798,158.28508966148235,154.51823846028506,155.29376664876685,154.52377794734562,169.72136069805862,194.06186684226594,160.43441064098909,190.29778538459891,183.92183577786642,158.32663581443674,156.60939482565564,150.35254419072567,191.59956484383622,154.02245436836276,175.24145955393087,172.86224986140991,155.39624715938766,171.21702220441637,155.58735946297782,165.79940385916493,153.69839437531857,172.8677893484705,155.39901690291794,175.59875646933855,164.09878133156556,194.77923041661163,164.07385363979293,164.07939312685352,158.03858248728636,179.09140306103694,181.70881069716302,188.11799722625901,158.04135223081664,155.44333279940264,158.28785940501265,150.00355650590885,152.5240231184747,178.25494051488869,190.13991000337231,149.9592406094242,168.67993713066878,190.26731820576572,152.50740465729297,155.90587996896141,153.73717078474266,152.00054159124949,155.41009587703911,154.82844973567776,166.58047153470733,170.54674427008567,177.06395079686308,191.34751818257962,194.44409144944626,190.26454846223544,151.68756057232648,153.69562463178829,154.02799385542335,190.15929820808429,150.12542522124173,154.06123077778685,167.46124997734023,172.72376268489529,161.60878189783293,167.58588843620339,179.79768766126142,194.78476990367221,158.00534556492286,170.58829042304004,165.06265208010723,170.57444170538858,151.11976314861658,155.65660305123512,154.04738206013539,182.49264811623567,173.32479703096868,150.82894007793593,150.15312265654464,183.36788707180798,155.64275433358367,172.77361806844056,163.01027212416076,158.36818196739111,192.98997609604288,154.54316615205767,158.39587940269405,151.9950021041889,150.97850622857169,150.55196572490669,150.64613700493663,174.11417393710195,176.21917902012396,154.19971795430146,163.05181827711516,169.12586583904582,153.70947334943975,194.78753964720249,158.36541222386083,152.83423439386743,191.56355817794241],"y":[8.5261105650160811,7.7747573289957703,7.7747573289957703,7.3967875988779657,6.8544298806121544,6.7066699712236986,6.6892798262804725,6.5057726928370645,6.4445325991417235,6.4073485153010656,6.2927051179793736,6.2302525151489778,6.2030168053539594,6.1188023601820118,6.064575305111326,5.8850345415346821,5.8765114074657365,5.7959739811080739,5.590614872686074,5.4425647120957406,5.1795956568055557,4.9727344252367995,4.8520143151587467,4.8483056012847268,4.7580139472546525,4.7170008448539393,4.6765644449688271,4.6098943342490859,4.6098943342490859,4.6052858089140001,4.5001715000864753,4.4717127522435387,4.4305367593250127,4.3769497925293788,4.3738423868296037,4.3400857853349963,4.3339237707174876,4.2698377866468524,4.2369356606541499,4.2174342080712375,4.168650077369767,4.1263065115912534,4.0771098213532015,4.0771098213532015,4.0757308400634535,4.0649726883193651,4.0464770623340591,4.0144542842572939,4.0031575444027956,3.9696508479789649,3.9646186852793996,3.9046131707086635,3.8955802021792989,3.8918108760748735,3.8394163874904597,3.7285855540687196,3.6856744292359025,3.6256361484139776,3.5806503469187363,3.5613479361987528,3.5278216315904389,3.4330571054794841,3.3388658889849632,3.3111839187309768,3.215616314172749,3.210157760739563,3.1678977980798426,3.1575287262804492,3.1320547394057883,3.0666399146971899,3.0599299604217123,3.0400885368514077,2.9258438041950972,2.9189497979157681,2.8659454439925942,2.8609615626523746,2.8586169363879876,2.8493968075637497,2.8493968075637497,2.7993514240567792,2.7910425717363205,2.7849553655062693,2.7849553655062693,2.7105339876534598,2.6966216366512823,2.692873422475337,2.689361256338108,2.6850667316165673,2.6808969576238155,2.6801629545087819,2.6657684137864859,2.6261339141596505,2.6175471758791593,2.6104027615740106,2.5581583379650428,2.4968889175468405,2.4734205653735488,2.415414881063906,2.3486824164600222,2.333946493119976,2.3064282639870788,2.3023901777117222,2.2795413424199586,2.2795413424199586,2.1616858761729363,2.1607127449634222,2.1407906564783148,2.1248826037657653,2.0918408667313346,2.0103881518193862,1.9673034836179901,1.9650561799930077,1.9380257977751645,1.8812407535075162,1.8786271366642688,1.8764088247576265,1.8625362017105203,1.8281238752396505,1.8166941728891759,1.78424509954147,1.7683954330620451,1.7652657019918547,1.6895162791935359,1.6437867756896338,1.6419679359516517,1.6081879625897038,1.6081879625897038,1.6044050582011367,1.6028445509905489,1.5799625526317507,1.5799625526317507,1.5729129514736235,1.5590074546223163,1.5153889944591026,1.503804162080469,1.503804162080469,1.4832519778219835,1.4684855544997726,1.4627555480184777,1.4462140027399637,1.4354892200434339,1.4354892200434339,1.4005581738810804,1.3945120256429016,1.368258880151469,1.3564012955823259,1.3235797452658071,1.3078038762449913,1.3075480022592016,1.3025401195616002],"text":["HP:0004377 (236) <br> Hematological neoplasm <br> 2.978e-09","HP:0012848 (24) <br> Small intestinal stenosis <br> 1.680e-08","HP:0100867 (24) <br> Duodenal stenosis <br> 1.680e-08","HP:0000340 (164) <br> Sloping forehead <br> 4.011e-08","HP:0005522 (23) <br> Pyridoxine-responsive sideroblastic anemia <br> 1.398e-07","HP:0003220 (49) <br> Abnormality of chromosome stability <br> 1.965e-07","HP:0031105 (99) <br> Abnormal uterus morphology <br> 2.045e-07","HP:0010469 (24) <br> Absent testis <br> 3.121e-07","HP:0100760 (28) <br> Clubbing of toes <br> 3.593e-07","HP:0003022 (41) <br> Hypoplasia of the ulna <br> 3.914e-07","HP:0032101 (763) <br> Unusual infection <br> 5.097e-07","HP:0008572 (62) <br> External ear malformation <br> 5.885e-07","HP:0000813 (33) <br> Bicornuate uterus <br> 6.266e-07","HP:0002664 (691) <br> Neoplasm <br> 7.607e-07","HP:0005368 (241) <br> Abnormality of humoral immunity <br> 8.618e-07","HP:0011793 (669) <br> Neoplasm by anatomical site <br> 1.303e-06","HP:0002245 (26) <br> Meckel diverticulum <br> 1.329e-06","HP:0006495 (44) <br> Aplasia/Hypoplasia of the ulna <br> 1.600e-06","HP:0010293 (27) <br> Aplasia/Hypoplasia of the uvula <br> 2.567e-06","HP:0002719 (711) <br> Recurrent infections <br> 3.609e-06","HP:0009821 (58) <br> Forearm undergrowth <br> 6.613e-06","HP:0012252 (1008) <br> Abnormal respiratory system morphology <br> 1.065e-05","HP:0000010 (110) <br> Recurrent urinary tract infections <br> 1.406e-05","HP:0009824 (66) <br> Upper limb undergrowth <br> 1.418e-05","HP:0002778 (111) <br> Abnormal tracheal morphology <br> 1.746e-05","HP:0000027 (67) <br> Azoospermia <br> 1.919e-05","HP:0025033 (775) <br> Abnormality of digestive system morphology <br> 2.106e-05","HP:0005344 (62) <br> Abnormal carotid artery morphology <br> 2.455e-05","HP:0002246 (62) <br> Abnormal duodenum morphology <br> 2.455e-05","HP:0008669 (80) <br> Abnormal spermatogenesis <br> 2.481e-05","HP:0012718 (771) <br> Morphological abnormality of the gastrointestinal tract <br> 3.161e-05","HP:0011893 (390) <br> Abnormal leukocyte count <br> 3.375e-05","HP:0010460 (382) <br> Abnormality of the female genitalia <br> 3.711e-05","HP:0001924 (32) <br> Sideroblastic anemia <br> 4.198e-05","HP:0007565 (58) <br> Multiple cafe-au-lait spots <br> 4.228e-05","HP:0006503 (70) <br> Aplasia/hypoplasia involving forearm bones <br> 4.570e-05","HP:0007874 (42) <br> Almond-shaped palpebral fissure <br> 4.635e-05","HP:0000130 (160) <br> Abnormality of the uterus <br> 5.372e-05","HP:0006501 (59) <br> Aplasia/Hypoplasia of the radius <br> 5.795e-05","HP:0002575 (65) <br> Tracheoesophageal fistula <br> 6.061e-05","HP:0001549 (33) <br> Abnormal ileum morphology <br> 6.782e-05","HP:0000481 (472) <br> Abnormal cornea morphology <br> 7.476e-05","HP:0001881 (511) <br> Abnormal leukocyte morphology <br> 8.373e-05","HP:0010987 (511) <br> Abnormal cellular immune system morphology <br> 8.373e-05","HP:0005372 (226) <br> Abnormality of B cell physiology <br> 8.400e-05","HP:0010978 (1203) <br> Abnormality of immune system physiology <br> 8.610e-05","HP:0001824 (243) <br> Weight loss <br> 8.985e-05","HP:0002863 (92) <br> Myelodysplasia <br> 9.673e-05","HP:0000520 (195) <br> Proptosis <br> 9.928e-05","HP:0100587 (34) <br> Abnormal preputium morphology <br> 1.072e-04","HP:0001871 (1101) <br> Abnormality of blood and blood-forming tissues <br> 1.085e-04","HP:0004313 (173) <br> Decreased circulating antibody level <br> 1.246e-04","HP:0002242 (603) <br> Abnormal intestine morphology <br> 1.272e-04","HP:0002665 (100) <br> Lymphoma <br> 1.283e-04","HP:0002244 (174) <br> Abnormality of the small intestine <br> 1.447e-04","HP:0010701 (224) <br> Abnormal immunoglobulin level <br> 1.868e-04","HP:0200005 (46) <br> Abnormal shape of the palpebral fissure <br> 2.062e-04","HP:0005607 (251) <br> Abnormal tracheobronchial morphology <br> 2.368e-04","HP:0040088 (186) <br> Abnormal lymphocyte count <br> 2.626e-04","HP:0032251 (526) <br> Abnormal immune system morphology <br> 2.746e-04","HP:0004332 (203) <br> Abnormal lymphocyte morphology <br> 2.966e-04","HP:0003330 (559) <br> Abnormal bone structure <br> 3.689e-04","HP:0000172 (143) <br> Abnormal uvula morphology <br> 4.583e-04","HP:0100026 (43) <br> Arteriovenous malformation <br> 4.884e-04","HP:0002012 (1017) <br> Abnormality of the abdominal organs <br> 6.087e-04","HP:0012745 (87) <br> Short palpebral fissure <br> 6.164e-04","HP:0011873 (340) <br> Abnormal platelet count <br> 6.794e-04","HP:0002715 (1443) <br> Abnormality of the immune system <br> 6.958e-04","HP:0011267 (9) <br> Microtia, third degree <br> 7.378e-04","HP:0002818 (147) <br> Abnormal morphology of the radius <br> 8.577e-04","HP:0009179 (280) <br> Deviation of the 5th finger <br> 8.711e-04","HP:0001873 (316) <br> Thrombocytopenia <br> 9.118e-04","HP:0011875 (346) <br> Abnormal platelet morphology <br> 1.186e-03","HP:0002716 (173) <br> Lymphadenopathy <br> 1.205e-03","HP:0012874 (182) <br> Abnormal male reproductive system physiology <br> 1.362e-03","HP:0008062 (105) <br> Aplasia/Hypoplasia affecting the anterior segment of the eye <br> 1.377e-03","HP:0410240 (84) <br> Abnormal circulating IgA level <br> 1.385e-03","HP:0008053 (98) <br> Aplasia/Hypoplasia of the iris <br> 1.415e-03","HP:0008055 (98) <br> Aplasia/Hypoplasia affecting the uvea <br> 1.415e-03","HP:0004207 (377) <br> Abnormal 5th finger morphology <br> 1.587e-03","HP:0030406 (13) <br> Primary peritoneal carcinoma <br> 1.618e-03","HP:0031409 (251) <br> Abnormal lymphocyte physiology <br> 1.641e-03","HP:0033796 (251) <br> Abnormal leukocyte physiology <br> 1.641e-03","HP:0004209 (279) <br> Clinodactyly of the 5th finger <br> 1.947e-03","HP:0002733 (185) <br> Abnormal lymph node morphology <br> 2.011e-03","HP:0004315 (79) <br> Decreased circulating IgG level <br> 2.028e-03","HP:0000025 (145) <br> Functional abnormality of male internal genitalia <br> 2.045e-03","HP:0001217 (100) <br> Clubbing <br> 2.065e-03","HP:0030084 (399) <br> Clinodactyly <br> 2.085e-03","HP:0040012 (31) <br> Chromosome breakage <br> 2.089e-03","HP:0000008 (316) <br> Abnormal morphology of female internal genitalia <br> 2.159e-03","HP:0010301 (108) <br> Spinal dysraphism <br> 2.365e-03","HP:0040072 (162) <br> Abnormal forearm bone morphology <br> 2.412e-03","HP:0001199 (48) <br> Triphalangeal thumb <br> 2.452e-03","HP:0002973 (163) <br> Abnormal forearm morphology <br> 2.766e-03","HP:0001888 (156) <br> Lymphopenia <br> 3.185e-03","HP:0000953 (249) <br> Hyperpigmentation of the skin <br> 3.362e-03","HP:0002720 (62) <br> Decreased circulating IgA level <br> 3.842e-03","HP:0002414 (97) <br> Spina bifida <br> 4.480e-03","HP:0009484 (371) <br> Deviation of the hand or of fingers of the hand <br> 4.635e-03","HP:0011017 (517) <br> Abnormal cellular physiology <br> 4.938e-03","HP:0025354 (537) <br> Abnormal cellular phenotype <br> 4.984e-03","HP:0045005 (227) <br> Neural tube defect <br> 5.254e-03","HP:0410043 (227) <br> Abnormal neural tube morphology <br> 5.254e-03","HP:0040071 (122) <br> Abnormal morphology of ulna <br> 6.892e-03","HP:0000812 (357) <br> Abnormal internal genitalia <br> 6.907e-03","HP:0001872 (367) <br> Abnormality of thrombocytes <br> 7.231e-03","HP:0002014 (293) <br> Diarrhea <br> 7.501e-03","HP:0040019 (303) <br> Finger clinodactyly <br> 8.094e-03","HP:0000072 (53) <br> Hydroureter <br> 9.764e-03","HP:0002031 (157) <br> Abnormal esophagus morphology <br> 1.078e-02","HP:0009826 (182) <br> Limb undergrowth <br> 1.084e-02","HP:0011821 (980) <br> Abnormal facial skeleton morphology <br> 1.153e-02","HP:0006443 (16) <br> Patellar aplasia <br> 1.314e-02","HP:0009892 (12) <br> Anotia <br> 1.322e-02","HP:0030669 (1147) <br> Abnormal ocular adnexa morphology <br> 1.329e-02","HP:0410242 (96) <br> Abnormal circulating IgG level <br> 1.372e-02","HP:0004097 (357) <br> Deviation of finger <br> 1.486e-02","HP:0011032 (494) <br> Abnormality of fluid regulation <br> 1.525e-02","HP:0008678 (238) <br> Renal hypoplasia/aplasia <br> 1.643e-02","HP:0011027 (21) <br> Abnormal fallopian tube morphology <br> 1.705e-02","HP:0000553 (378) <br> Abnormal uvea morphology <br> 1.717e-02","HP:0002850 (56) <br> Decreased circulating total IgM <br> 2.044e-02","HP:0002023 (99) <br> Anal atresia <br> 2.271e-02","HP:0031703 (1182) <br> Abnormal ear morphology <br> 2.281e-02","HP:0012041 (92) <br> Decreased fertility in males <br> 2.465e-02","HP:0000415 (92) <br> Abnormality of the choanae <br> 2.465e-02","HP:0000083 (288) <br> Renal insufficiency <br> 2.487e-02","HP:0032039 (1162) <br> Abnormality of the ocular adnexa <br> 2.495e-02","HP:0002843 (85) <br> Abnormal T cell morphology <br> 2.630e-02","HP:0011839 (85) <br> Abnormal T cell count <br> 2.630e-02","HP:0007378 (234) <br> Neoplasm of the gastrointestinal tract <br> 2.674e-02","HP:0004349 (365) <br> Reduced bone mineral density <br> 2.761e-02","HP:0100542 (157) <br> Abnormal localization of kidney <br> 3.052e-02","HP:0002251 (101) <br> Aganglionic megacolon <br> 3.135e-02","HP:0004362 (101) <br> Abnormality of enteric ganglion morphology <br> 3.135e-02","HP:0000951 (1484) <br> Abnormality of the skin <br> 3.287e-02","HP:0000492 (993) <br> Abnormal eyelid morphology <br> 3.400e-02","HP:0000268 (158) <br> Dolichocephaly <br> 3.445e-02","HP:0000315 (1182) <br> Abnormality of the orbital region <br> 3.579e-02","HP:0012331 (102) <br> Abnormal autonomic nervous system morphology <br> 3.669e-02","HP:0025028 (102) <br> Abnormality of enteric nervous system morphology <br> 3.669e-02","HP:0002088 (866) <br> Abnormal lung morphology <br> 3.976e-02","HP:0007400 (185) <br> Irregular hyperpigmentation <br> 4.032e-02","HP:0010468 (103) <br> Aplasia/Hypoplasia of the testes <br> 4.283e-02","HP:0001877 (531) <br> Abnormal erythrocyte morphology <br> 4.401e-02","HP:0410243 (81) <br> Abnormal circulating IgM level <br> 4.747e-02","HP:0004348 (422) <br> Abnormality of bone mineral density <br> 4.923e-02","HP:0001392 (849) <br> Abnormality of the liver <br> 4.926e-02","HP:0100013 (74) <br> Neoplasm of the breast <br> 4.983e-02"],"key":["HP:0004377","HP:0012848","HP:0100867","HP:0000340","HP:0005522","HP:0003220","HP:0031105","HP:0010469","HP:0100760","HP:0003022","HP:0032101","HP:0008572","HP:0000813","HP:0002664","HP:0005368","HP:0011793","HP:0002245","HP:0006495","HP:0010293","HP:0002719","HP:0009821","HP:0012252","HP:0000010","HP:0009824","HP:0002778","HP:0000027","HP:0025033","HP:0005344","HP:0002246","HP:0008669","HP:0012718","HP:0011893","HP:0010460","HP:0001924","HP:0007565","HP:0006503","HP:0007874","HP:0000130","HP:0006501","HP:0002575","HP:0001549","HP:0000481","HP:0001881","HP:0010987","HP:0005372","HP:0010978","HP:0001824","HP:0002863","HP:0000520","HP:0100587","HP:0001871","HP:0004313","HP:0002242","HP:0002665","HP:0002244","HP:0010701","HP:0200005","HP:0005607","HP:0040088","HP:0032251","HP:0004332","HP:0003330","HP:0000172","HP:0100026","HP:0002012","HP:0012745","HP:0011873","HP:0002715","HP:0011267","HP:0002818","HP:0009179","HP:0001873","HP:0011875","HP:0002716","HP:0012874","HP:0008062","HP:0410240","HP:0008053","HP:0008055","HP:0004207","HP:0030406","HP:0031409","HP:0033796","HP:0004209","HP:0002733","HP:0004315","HP:0000025","HP:0001217","HP:0030084","HP:0040012","HP:0000008","HP:0010301","HP:0040072","HP:0001199","HP:0002973","HP:0001888","HP:0000953","HP:0002720","HP:0002414","HP:0009484","HP:0011017","HP:0025354","HP:0045005","HP:0410043","HP:0040071","HP:0000812","HP:0001872","HP:0002014","HP:0040019","HP:0000072","HP:0002031","HP:0009826","HP:0011821","HP:0006443","HP:0009892","HP:0030669","HP:0410242","HP:0004097","HP:0011032","HP:0008678","HP:0011027","HP:0000553","HP:0002850","HP:0002023","HP:0031703","HP:0012041","HP:0000415","HP:0000083","HP:0032039","HP:0002843","HP:0011839","HP:0007378","HP:0004349","HP:0100542","HP:0002251","HP:0004362","HP:0000951","HP:0000492","HP:0000268","HP:0000315","HP:0012331","HP:0025028","HP:0002088","HP:0007400","HP:0010468","HP:0001877","HP:0410243","HP:0004348","HP:0001392","HP:0100013"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(153,0,153,1)","opacity":0.80000000000000004,"size":[17.242706764704529,12.704122251394768,12.704122251394768,16.627795233541928,12.597492423457624,14.333066067312378,15.722754841180334,12.704122251394768,13.080107699525154,13.950740632328149,19.057585561962338,14.817530145783991,13.464799414642561,18.912702804588378,17.277273632652467,18.865106592355996,12.901287144618975,14.103902153872516,12.992787250079516,18.954551180399363,14.682412279994908,19.457460196822758,15.917246790822361,14.94271233506351,15.933807339693688,14.972613250113406,19.080270654784396,14.817530145783991,14.817530145783991,15.319377109855813,19.072751951243912,18.047898374760198,18.015575995116624,13.393923550484519,14.682412279994908,15.059257534741903,14.003263398683421,16.585008068656709,14.717204984820066,14.912272742159212,13.464799414642561,18.342130271142743,18.462800953621461,18.462800953621461,17.171037487721456,19.706204183578347,17.290875682983636,15.585496862893729,16.923887654084208,13.533067077450161,19.58207626128576,16.719885765004154,18.711290391323303,15.741443860532286,16.729781696869093,17.156276088848124,14.199158443904556,17.344055370798454,16.843741584418417,18.506530098673746,16.991699324542299,18.598079039514069,16.388532487449179,14.05429113859868,19.470056827841098,15.479821521486015,17.832550627610072,19.958069469546835,9.7835234899655621,16.43706926513433,17.522042437262439,17.716290984254197,17.860190679978924,16.719885765004154,16.806702574618992,15.831759381153528,15.412969781594908,15.703846343452833,15.703846343452833,17.994990730367135,11.0179178849601,17.344055370798454,17.344055370798454,17.516254668148232,16.834566846316818,15.295124283351443,16.412991558022448,15.741443860532286,18.083393392381662,13.32024543969502,17.716290984254197,15.883599641936872,16.606551553606309,14.289551053282345,16.617210370299976,16.540988618015533,17.330940472631156,14.817530145783991,15.684713413901202,17.969884649479432,18.48046037198165,18.537733218257774,17.17836322734254,17.17836322734254,16.105389877305754,17.909506598508575,17.952895489455628,17.595247170324892,17.649132701483286,14.497057283970744,16.552112715628681,16.806702574618992,19.417471883690116,11.629649013742458,10.767816834397534,19.639541257708668,15.66535093296908,17.909506598508575,18.4114961423594,17.256631257038578,12.365130105867371,17.999131931548465,14.610639892328045,15.722754841180334,19.681609019006014,15.585496862893729,15.585496862893729,17.567529019446152,19.657739024278342,15.435558208980297,15.435558208980297,17.228649163403773,17.944323619431238,16.552112715628681,15.759918195883458,15.759918195883458,19.996513752768735,19.436190784089451,16.5631565955727,19.681609019006014,15.7781824892029,15.7781824892029,19.240703190994726,16.834566846316818,15.796241236378519,18.52080163770313,15.34327872942024,18.170214667504137,19.21217628704002,15.168229566608264],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(153,0,153,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"HP","legendgroup":"HP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[130.37958479804664,130.25473196212985,130.31299661889102,130.31854563382066,130.31022211142621,130.85957458946007,130.56825130565423,129.69705596170158,129.66376187212376,130.88177064917861,130.99552545523613,130.30744760396138,130.57657482804871,130.39345733537073,130.98165291791204,130.30467309649657,130.57934933551351,130.98442742537685,130.9927509477713,130.39068282790592,129.97728121564813,130.9761039029824,130.73472175354328,130.98720193284166,130.51831017128754,129.72757554381457,130.59877088776724,130.97887841044721],"y":[17,17,9.042040388506793,9.0094312811076414,7.4395290787643544,5.6621872042386974,5.2390400221311317,5.0565494469989281,5.0216209455211676,4.2600259266374287,3.9987146068508457,3.6386462054072206,3.2181981560009891,3.1471916570688898,2.8730137640379629,2.6459234378038952,2.6402546908890963,2.6172441899845591,2.5069357283082776,1.8814808073615021,1.7731307188597119,1.6140620844283744,1.5212220595530304,1.5081679823094152,1.452862792244948,1.3852465153877904,1.3712211768481086,1.3712211768481086],"text":["KEGG:04110 (150) <br> Cell cycle <br> 4.050e-22","KEGG:03030 (35) <br> DNA replication <br> 8.819e-19","KEGG:03440 (39) <br> Homologous recombination <br> 9.077e-10","KEGG:03460 (48) <br> Fanconi anemia pathway <br> 9.785e-10","KEGG:03430 (21) <br> Mismatch repair <br> 3.635e-08","KEGG:05150 (36) <br> Staphylococcus aureus infection <br> 2.177e-06","KEGG:04640 (72) <br> Hematopoietic cell lineage <br> 5.767e-06","KEGG:00100 (17) <br> Steroid biosynthesis <br> 8.779e-06","KEGG:00000 (5291) <br> KEGG root term <br> 9.514e-06","KEGG:05166 (202) <br> Human T-cell leukemia virus 1 infection <br> 5.495e-05","KEGG:05340 (30) <br> Primary immunodeficiency <br> 1.003e-04","KEGG:03420 (43) <br> Nucleotide excision repair <br> 2.298e-04","KEGG:04658 (72) <br> Th1 and Th2 cell differentiation <br> 6.051e-04","KEGG:04115 (66) <br> p53 signaling pathway <br> 7.125e-04","KEGG:05321 (43) <br> Inflammatory bowel disease <br> 1.340e-03","KEGG:03410 (33) <br> Base excision repair <br> 2.260e-03","KEGG:04659 (86) <br> Th17 cell differentiation <br> 2.290e-03","KEGG:05322 (79) <br> Systemic lupus erythematosus <br> 2.414e-03","KEGG:05332 (34) <br> Graft-versus-host disease <br> 3.112e-03","KEGG:04114 (97) <br> Oocyte meiosis <br> 1.314e-02","KEGG:00670 (17) <br> One carbon pool by folate <br> 1.686e-02","KEGG:05310 (13) <br> Asthma <br> 2.432e-02","KEGG:04940 (36) <br> Type I diabetes mellitus <br> 3.011e-02","KEGG:05323 (64) <br> Rheumatoid arthritis <br> 3.103e-02","KEGG:04514 (88) <br> Cell adhesion molecules <br> 3.525e-02","KEGG:00240 (44) <br> Pyrimidine metabolism <br> 4.119e-02","KEGG:04672 (31) <br> Intestinal immune network for IgA production <br> 4.254e-02","KEGG:05320 (31) <br> Autoimmune thyroid disease <br> 4.254e-02"],"key":["KEGG:04110","KEGG:03030","KEGG:03440","KEGG:03460","KEGG:03430","KEGG:05150","KEGG:04640","KEGG:00100","KEGG:00000","KEGG:05166","KEGG:05340","KEGG:03420","KEGG:04658","KEGG:04115","KEGG:05321","KEGG:03410","KEGG:04659","KEGG:05322","KEGG:05332","KEGG:04114","KEGG:00670","KEGG:05310","KEGG:04940","KEGG:05323","KEGG:04514","KEGG:00240","KEGG:04672","KEGG:05320"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(221,68,119,1)","opacity":0.80000000000000004,"size":[16.472495106982581,13.59890151556252,13.840863529293893,14.289551053282345,12.365130105867371,13.662461030030807,15.114631326484991,11.799460832482184,21.654139222746778,16.98338923803659,13.243549133997096,14.05429113859868,15.114631326484991,14.94271233506351,14.05429113859868,13.464799414642561,15.457839510796521,15.295124283351443,13.533067077450161,15.684713413901202,11.799460832482184,11.0179178849601,13.662461030030807,14.88127578654996,15.501511785319126,14.103902153872516,13.32024543969502,13.32024543969502],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(221,68,119,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"KEGG","legendgroup":"KEGG","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[146.3168232200351],"y":[2.000242846070146],"text":"MIRNA:mmu-miR-5101 (139) <br> mmu-miR-5101 <br> 9.994e-03","key":["MIRNA:mmu-miR-5101"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(34,170,153,1)","opacity":0.80000000000000004,"size":16.338422958806614,"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(34,170,153,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"MIRNA","legendgroup":"MIRNA","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isSimpleKey":true,"_isNestedKey":false,"frame":null},{"x":[134.68264136794994,134.68818375279321,134.68541256037159,136.36752636029541,136.13197500445764,137.55636790917075,135.13434573267415,134.27804727439332,134.28081846681494,136.37306874513865,137.11297712171142,136.36198397545218,136.35367039818729,134.18936911690145,134.76577714059857,135.69135540941991,134.11731811393932,135.75509283511721,134.95698941769041,135.69412660184156,137.10743473686819,137.7752921104788,134.95144703284717,137.6034781803383,135.49182955506325,138.01084346631657,134.95421822526879,138.75075184288931,134.96530299495529,137.90830934671658,137.91108053913823,136.35644159060894,135.01241326612282,136.61139129339804,135.48628717022001,135.75232164269559,136.09317831055495,136.92176484461959,136.08763592571171,136.98827346273848,136.977188693052,137.55082552432751,137.55359671674913,134.74360760122562,135.49460074748487,136.06546638633873,134.95976061011203,135.25350700680386,136.70006945088991,138.2685643615273,137.33744370786269,138.26025078426244,136.59199294644671,135.6996689866848,136.92453603704121,137.54251194706265,137.45106259714916,138.17988620403543,135.57496532771185,138.27687793879218,134.93759107073907,135.82437264565775,134.4082933182095,136.37029755271703,134.05912307308529,135.50014313232811,136.59476413886833,136.35921278303056,136.26222104827383,135.49737193990646,137.54805433190589,136.99935823242495,137.53142717737617,137.37069801692215,137.79469045743014,137.69492753025179,138.19374216614355,138.4403782916678,137.41226590324646,135.3421851642957,137.34298609270596,136.94393438399254,136.87465457345203,134.80180264207965,134.88493841472825,134.28636085165817,137.98867392694362,136.312102511863,136.30656012701976,136.91899365219797,137.43997782746266,136.60030652371157,134.87108245262016,134.96253180253365,136.11534784992793,136.11811904234955,137.34575728512758,134.72975163911752,137.37901159418701,138.44314948408942],"y":[17,17,17,17,17,17,17,17,17,17,17,17,17,17,14.383013083154358,14.127036851331852,13.724854735128746,13.683553953142219,13.373133585759694,13.315297453081808,12.640457995736629,12.174080921317199,11.872185864307154,10.859249648876283,10.625514266439206,10.384269481355995,9.4836379693985808,9.32480295232053,9.1734775975612575,7.7297699281208754,7.5443260931564664,7.3248603928633687,7.2957520079430909,7.2957520079430909,7.1713880450270073,7.1043161104421904,6.8333080501535486,6.8333080501535486,6.7023161370744795,6.6743397187248314,6.5442749340581425,6.4890702031425338,6.4890702031425338,6.4750548250708615,6.0913698958724494,5.8856885285035263,5.6365889536090306,5.2922343081720076,5.267805376680772,5.1090710664366954,5.0955775059844743,5.0955775059844743,4.9723858044347296,4.7671934949748929,4.2061577163979651,4.1966446129692567,4.0424425655674039,4.0136436451428494,3.902875778092342,3.851816314244715,3.7601808191841255,3.7421767769470362,3.6509808245464059,3.6244046911856009,3.377812384955774,3.1567017555701677,3.1001623867963071,3.0397711809141428,3.032365242411009,2.986681704657872,2.9081923533534035,2.8277508307822843,2.8277508307822843,2.7823946605148788,2.7485308340955537,2.7238931284873553,2.7179910851631224,2.6310763146930936,2.5846466759567046,2.4610796193156457,2.4115057787023555,2.3669388506671702,2.2368777925063084,2.2368777925063084,2.1687794360272714,2.0264407478153066,1.952478964721299,1.901039896997093,1.901039896997093,1.8668407888644669,1.8351271151622177,1.727500789871242,1.7211022605739761,1.7211022605739761,1.7182790978991611,1.7182790978991611,1.642410921883968,1.642410921883968,1.4305698925784303,1.3155227381947396],"text":["REAC:R-MMU-1640170 (524) <br> Cell Cycle <br> 1.926e-52","REAC:R-MMU-69278 (446) <br> Cell Cycle, Mitotic <br> 1.277e-47","REAC:R-MMU-69620 (248) <br> Cell Cycle Checkpoints <br> 1.170e-31","REAC:R-MMU-68877 (169) <br> Mitotic Prometaphase <br> 2.117e-25","REAC:R-MMU-68886 (333) <br> M Phase <br> 8.489e-23","REAC:R-MMU-2500257 (105) <br> Resolution of Sister Chromatid Cohesion <br> 1.094e-21","REAC:R-MMU-9648025 (93) <br> EML4 and NUDC in mitotic spindle formation <br> 2.541e-20","REAC:R-MMU-141444 (84) <br> Amplification  of signal from unattached  kinetochores via a MAD2  inhibitory signal <br> 2.537e-19","REAC:R-MMU-141424 (84) <br> Amplification of signal from the kinetochores <br> 2.537e-19","REAC:R-MMU-69618 (100) <br> Mitotic Spindle Checkpoint <br> 8.023e-19","REAC:R-MMU-5663220 (118) <br> RHO GTPases Activate Formins <br> 3.238e-18","REAC:R-MMU-2555396 (206) <br> Mitotic Metaphase and Anaphase <br> 8.569e-18","REAC:R-MMU-68882 (205) <br> Mitotic Anaphase <br> 3.421e-17","REAC:R-MMU-68962 (29) <br> Activation of the pre-replicative complex <br> 7.771e-17","REAC:R-MMU-73886 (95) <br> Chromosome Maintenance <br> 4.140e-15","REAC:R-MMU-5685942 (51) <br> HDR through Homologous Recombination (HRR) <br> 7.464e-15","REAC:R-MMU-176187 (34) <br> Activation of ATR in response to replication stress <br> 1.884e-14","REAC:R-MMU-5693538 (99) <br> Homology Directed Repair <br> 2.072e-14","REAC:R-MMU-69306 (117) <br> DNA Replication <br> 4.235e-14","REAC:R-MMU-5693567 (96) <br> HDR through Homologous Recombination (HRR) or Single Strand Annealing (SSA) <br> 4.838e-14","REAC:R-MMU-195258 (209) <br> RHO GTPase Effectors <br> 2.288e-13","REAC:R-MMU-2467813 (168) <br> Separation of Sister Chromatids <br> 6.698e-13","REAC:R-MMU-5693532 (122) <br> DNA Double-Strand Break Repair <br> 1.342e-12","REAC:R-MMU-69242 (136) <br> S Phase <br> 1.383e-11","REAC:R-MMU-69481 (138) <br> G2/M Checkpoints <br> 2.369e-11","REAC:R-MMU-69239 (110) <br> Synthesis of DNA <br> 4.128e-11","REAC:R-MMU-73894 (246) <br> DNA Repair <br> 3.284e-10","REAC:0000000 (5890) <br> REACTOME root term <br> 4.734e-10","REAC:R-MMU-69190 (19) <br> DNA strand elongation <br> 6.707e-10","REAC:R-MMU-194315 (481) <br> Signaling by Rho GTPases <br> 1.863e-08","REAC:R-MMU-9716542 (494) <br> Signaling by Rho GTPases, Miro GTPases and RHOBTB3 <br> 2.855e-08","REAC:R-MMU-453279 (124) <br> Mitotic G1 phase and G1/S transition <br> 4.733e-08","REAC:R-MMU-606279 (54) <br> Deposition of new CENPA-containing nucleosomes at the centromere <br> 5.061e-08","REAC:R-MMU-774815 (54) <br> Nucleosome assembly <br> 5.061e-08","REAC:R-MMU-69206 (100) <br> G1/S Transition <br> 6.739e-08","REAC:R-MMU-5693579 (24) <br> Homologous DNA Pairing and Strand Exchange <br> 7.865e-08","REAC:R-MMU-69109 (12) <br> Leading Strand Synthesis <br> 1.468e-07","REAC:R-MMU-69091 (12) <br> Polymerase switching <br> 1.468e-07","REAC:R-MMU-69186 (15) <br> Lagging Strand Synthesis <br> 1.985e-07","REAC:R-MMU-5693607 (63) <br> Processing of DNA double-strand break ends <br> 2.117e-07","REAC:R-MMU-5693616 (22) <br> Presynaptic phase of homologous DNA pairing and strand exchange <br> 2.856e-07","REAC:R-MMU-5693537 (30) <br> Resolution of D-Loop Structures <br> 3.243e-07","REAC:R-MMU-5693568 (30) <br> Resolution of D-loop Structures through Holliday Junction Intermediates <br> 3.243e-07","REAC:R-MMU-191273 (26) <br> Cholesterol biosynthesis <br> 3.349e-07","REAC:R-MMU-69473 (67) <br> G2/M DNA damage checkpoint <br> 8.103e-07","REAC:R-MMU-983189 (37) <br> Kinesins <br> 1.301e-06","REAC:R-MMU-69002 (95) <br> DNA Replication Pre-Initiation <br> 2.309e-06","REAC:R-MMU-180786 (40) <br> Extension of Telomeres <br> 5.102e-06","REAC:R-MMU-5651801 (15) <br> PCNA-Dependent Long Patch Base Excision Repair <br> 5.398e-06","REAC:R-MMU-157579 (41) <br> Telomere Maintenance <br> 7.779e-06","REAC:R-MMU-110314 (27) <br> Recognition of DNA damage by PCNA-containing replication complex <br> 8.025e-06","REAC:R-MMU-174417 (27) <br> Telomere C-strand (Lagging Strand) Synthesis <br> 8.025e-06","REAC:R-MMU-2995410 (58) <br> Nuclear Envelope (NE) Reassembly <br> 1.066e-05","REAC:R-MMU-5685938 (33) <br> HDR through Single Strand Annealing (SSA) <br> 1.709e-05","REAC:R-MMU-174411 (22) <br> Polymerase switching on the C-strand of the telomere <br> 6.221e-05","REAC:R-MMU-110373 (18) <br> Resolution of AP sites via the multiple-nucleotide patch replacement pathway <br> 6.359e-05","REAC:R-MMU-6804756 (78) <br> Regulation of TP53 Activity through Phosphorylation <br> 9.069e-05","REAC:R-MMU-6791312 (32) <br> TP53 Regulates Transcription of Cell Cycle Genes <br> 9.691e-05","REAC:R-MMU-5696397 (19) <br> Gap-filling DNA repair synthesis and ligation in GG-NER <br> 1.251e-04","REAC:R-MMU-5656169 (28) <br> Termination of translesion DNA synthesis <br> 1.407e-04","REAC:R-MMU-73893 (44) <br> DNA Damage Bypass <br> 1.737e-04","REAC:R-MMU-198933 (68) <br> Immunoregulatory interactions between a Lymphoid and a non-Lymphoid cell <br> 1.811e-04","REAC:R-MMU-73884 (29) <br> Base Excision Repair <br> 2.234e-04","REAC:R-MMU-68875 (89) <br> Mitotic Prophase <br> 2.375e-04","REAC:R-MMU-8854518 (65) <br> AURKA Activation by TPX2 <br> 4.190e-04","REAC:R-MMU-69275 (163) <br> G2/M Transition <br> 6.971e-04","REAC:R-MMU-2980766 (43) <br> Nuclear Envelope Breakdown <br> 7.940e-04","REAC:R-MMU-453274 (165) <br> Mitotic G2-G2/M phases <br> 9.125e-04","REAC:R-MMU-8957322 (75) <br> Metabolism of steroids <br> 9.282e-04","REAC:R-MMU-69478 (5) <br> G2/M DNA replication checkpoint <br> 1.031e-03","REAC:R-MMU-73933 (28) <br> Resolution of Abasic Sites (AP sites) <br> 1.235e-03","REAC:R-MMU-69183 (11) <br> Processive synthesis on the lagging strand <br> 1.487e-03","REAC:R-MMU-69166 (11) <br> Removal of the Flap Intermediate <br> 1.487e-03","REAC:R-MMU-176408 (34) <br> Regulation of APC/C activators between G1/S and early anaphase <br> 1.650e-03","REAC:R-MMU-162582 (1432) <br> Signal Transduction <br> 1.784e-03","REAC:R-MMU-4615885 (40) <br> SUMOylation of DNA replication proteins <br> 1.888e-03","REAC:R-MMU-6804116 (8) <br> TP53 Regulates Transcription of Genes Involved in G1 Cell Cycle Arrest <br> 1.914e-03","REAC:R-MMU-110313 (35) <br> Translesion synthesis by Y family DNA polymerases bypasses lesions on DNA template <br> 2.338e-03","REAC:R-MMU-2565942 (80) <br> Regulation of PLK1 Activity at G2/M Transition <br> 2.602e-03","REAC:R-MMU-983231 (96) <br> Factors involved in megakaryocyte development and platelet production <br> 3.459e-03","REAC:R-MMU-380320 (75) <br> Recruitment of NuMA to mitotic centrosomes <br> 3.877e-03","REAC:R-MMU-9615933 (21) <br> Postmitotic nuclear pore complex (NPC) reformation <br> 4.296e-03","REAC:R-MMU-176417 (6) <br> Phosphorylation of Emi1 <br> 5.796e-03","REAC:R-MMU-173623 (6) <br> Classical antibody-mediated complement activation <br> 5.796e-03","REAC:R-MMU-69273 (22) <br> Cyclin A/B1/B2 associated events during G2/M transition <br> 6.780e-03","REAC:R-MMU-5620912 (87) <br> Anchoring of the basal body to the plasma membrane <br> 9.409e-03","REAC:R-MMU-69052 (88) <br> Switching of origins to a post-replicative state <br> 1.116e-02","REAC:R-MMU-5358565 (10) <br> Mismatch repair (MMR) directed by MSH2:MSH6 (MutSalpha) <br> 1.256e-02","REAC:R-MMU-5358508 (10) <br> Mismatch Repair <br> 1.256e-02","REAC:R-MMU-156711 (4) <br> Polo-like kinase mediated events <br> 1.359e-02","REAC:R-MMU-5633007 (129) <br> Regulation of TP53 Activity <br> 1.462e-02","REAC:R-MMU-3301854 (30) <br> Nuclear Pore Complex (NPC) Disassembly <br> 1.873e-02","REAC:R-MMU-166786 (7) <br> Creation of C4 and C2 activators <br> 1.901e-02","REAC:R-MMU-68952 (7) <br> DNA replication initiation <br> 1.901e-02","REAC:R-MMU-380259 (62) <br> Loss of Nlp from mitotic centrosomes <br> 1.913e-02","REAC:R-MMU-380284 (62) <br> Loss of proteins required for interphase microtubule organization from the centrosome <br> 1.913e-02","REAC:R-MMU-380270 (70) <br> Recruitment of mitotic centrosome proteins and complexes <br> 2.278e-02","REAC:R-MMU-380287 (70) <br> Centrosome maturation <br> 2.278e-02","REAC:R-MMU-977606 (21) <br> Regulation of Complement cascade <br> 3.710e-02","REAC:R-MMU-202430 (12) <br> Translocation of ZAP-70 to Immunological synapse <br> 4.836e-02"],"key":["REAC:R-MMU-1640170","REAC:R-MMU-69278","REAC:R-MMU-69620","REAC:R-MMU-68877","REAC:R-MMU-68886","REAC:R-MMU-2500257","REAC:R-MMU-9648025","REAC:R-MMU-141444","REAC:R-MMU-141424","REAC:R-MMU-69618","REAC:R-MMU-5663220","REAC:R-MMU-2555396","REAC:R-MMU-68882","REAC:R-MMU-68962","REAC:R-MMU-73886","REAC:R-MMU-5685942","REAC:R-MMU-176187","REAC:R-MMU-5693538","REAC:R-MMU-69306","REAC:R-MMU-5693567","REAC:R-MMU-195258","REAC:R-MMU-2467813","REAC:R-MMU-5693532","REAC:R-MMU-69242","REAC:R-MMU-69481","REAC:R-MMU-69239","REAC:R-MMU-73894","REAC:0000000","REAC:R-MMU-69190","REAC:R-MMU-194315","REAC:R-MMU-9716542","REAC:R-MMU-453279","REAC:R-MMU-606279","REAC:R-MMU-774815","REAC:R-MMU-69206","REAC:R-MMU-5693579","REAC:R-MMU-69109","REAC:R-MMU-69091","REAC:R-MMU-69186","REAC:R-MMU-5693607","REAC:R-MMU-5693616","REAC:R-MMU-5693537","REAC:R-MMU-5693568","REAC:R-MMU-191273","REAC:R-MMU-69473","REAC:R-MMU-983189","REAC:R-MMU-69002","REAC:R-MMU-180786","REAC:R-MMU-5651801","REAC:R-MMU-157579","REAC:R-MMU-110314","REAC:R-MMU-174417","REAC:R-MMU-2995410","REAC:R-MMU-5685938","REAC:R-MMU-174411","REAC:R-MMU-110373","REAC:R-MMU-6804756","REAC:R-MMU-6791312","REAC:R-MMU-5696397","REAC:R-MMU-5656169","REAC:R-MMU-73893","REAC:R-MMU-198933","REAC:R-MMU-73884","REAC:R-MMU-68875","REAC:R-MMU-8854518","REAC:R-MMU-69275","REAC:R-MMU-2980766","REAC:R-MMU-453274","REAC:R-MMU-8957322","REAC:R-MMU-69478","REAC:R-MMU-73933","REAC:R-MMU-69183","REAC:R-MMU-69166","REAC:R-MMU-176408","REAC:R-MMU-162582","REAC:R-MMU-4615885","REAC:R-MMU-6804116","REAC:R-MMU-110313","REAC:R-MMU-2565942","REAC:R-MMU-983231","REAC:R-MMU-380320","REAC:R-MMU-9615933","REAC:R-MMU-176417","REAC:R-MMU-173623","REAC:R-MMU-69273","REAC:R-MMU-5620912","REAC:R-MMU-69052","REAC:R-MMU-5358565","REAC:R-MMU-5358508","REAC:R-MMU-156711","REAC:R-MMU-5633007","REAC:R-MMU-3301854","REAC:R-MMU-166786","REAC:R-MMU-68952","REAC:R-MMU-380259","REAC:R-MMU-380284","REAC:R-MMU-380270","REAC:R-MMU-380287","REAC:R-MMU-977606","REAC:R-MMU-202430"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(51,102,204,1)","opacity":0.80000000000000004,"size":[18.50077955435771,18.255394068551965,17.324338667297848,16.679643677666306,17.79960972649998,15.831759381153528,15.605832335857878,15.412969781594908,15.412969781594908,15.741443860532286,16.045129496863122,17.016355390127128,17.008181919332767,13.163593405554227,15.645753607568084,14.416985747273221,13.533067077450161,15.722754841180334,16.029697700807695,15.66535093296908,17.040610116928963,16.669414345673637,16.105389877305754,16.29974880271676,16.32563830192576,15.917246790822361,17.311045057268586,21.787004578126478,12.102003142235581,18.370930211085224,18.4114961423594,16.13467646674961,14.535751690269857,14.535751690269857,15.741443860532286,12.704122251394768,10.767816834397534,10.767816834397534,11.444741251072838,14.849701816387062,12.484727576713514,13.243549133997096,13.243549133997096,12.901287144618975,14.972613250113406,13.723889277645902,15.645753607568084,13.896638027387656,11.444741251072838,13.950740632328149,12.992787250079516,12.992787250079516,14.682412279994908,13.464799414642561,12.484727576713514,11.956334427574175,15.270510325237122,13.393923550484519,12.102003142235581,13.080107699525154,14.103902153872516,15.001993266891569,13.163593405554227,15.522917570397238,14.912272742159212,16.617210370299976,14.05429113859868,16.638307116945004,15.194396009723683,6.9290217959811251,13.080107699525154,10.485368589139895,10.485368589139895,13.533067077450161,19.947553337222011,13.896638027387656,9.3304016728025392,13.59890151556252,15.319377109855813,15.66535093296908,15.194396009723683,12.365130105867371,8.0249937960183306,8.0249937960183306,12.484727576713514,15.479821521486015,15.501511785319126,10.161652391358336,10.161652391358336,3.7795275590551185,16.205587015350208,13.243549133997096,8.7671506498633249,8.7671506498633249,14.817530145783991,14.817530145783991,15.059257534741903,15.059257534741903,12.365130105867371,10.767816834397534],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(51,102,204,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"REAC","legendgroup":"REAC","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[199.75226112377777,199.7773119693388,199.78009539662332,199.81071309675343,199.44608412247695,199.57133835028182,199.86638164244448,200.0027695793876,199.84411422416807,199.59360576855826],"y":[13.962844202373024,10.815922428997059,6.1306927082854665,4.5934854886156078,4.0462004910504632,2.2332414037955179,2.1784100697994337,2.1668453556630327,1.8622558543359669,1.6871953740984877],"text":["WP:WP413 (59) <br> G1 to S cell cycle control <br> 1.089e-14","WP:WP4346 (46) <br> Cholesterol metabolism with Bloch and Kandutsch-Russell pathways <br> 1.528e-11","WP:WP103 (15) <br> Cholesterol biosynthesis <br> 7.401e-07","WP:WP1257 (9) <br> Mismatch repair <br> 2.550e-05","WP:000000 (3010) <br> WIKIPATHWAYS <br> 8.991e-05","WP:WP2902 (59) <br> p53 signaling <br> 5.845e-03","WP:WP1258 (12) <br> Homologous recombination <br> 6.631e-03","WP:WP5242 (82) <br> Comprehensive IL-17A signaling <br> 6.810e-03","WP:WP200 (9) <br> Complement activation, classical pathway <br> 1.373e-02","WP:WP435 (25) <br> One-carbon metabolism <br> 2.055e-02"],"key":["WP:WP413","WP:WP4346","WP:WP103","WP:WP1257","WP:000000","WP:WP2902","WP:WP1258","WP:WP5242","WP:WP200","WP:WP435"],"type":"scatter","mode":"markers","marker":{"autocolorscale":false,"color":"rgba(0,153,198,1)","opacity":0.80000000000000004,"size":[14.717204984820066,14.199158443904556,11.444741251072838,9.7835234899655621,20.938416852749633,14.717204984820066,10.767816834397534,15.366838671154913,9.7835234899655621,12.805215053656443],"symbol":"circle","line":{"width":1.8897637795275593,"color":"rgba(0,153,198,1)"}},"hoveron":"points","set":"SharedData9beb5615","name":"WP","legendgroup":"WP","showlegend":true,"xaxis":"x","yaxis":"y","hoverinfo":"text","_isNestedKey":false,"frame":null},{"x":[2,33.193772642710265],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(220,57,18,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[35.963352030325488,47.160761494453844],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(16,150,24,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[49.930340882069068,126.89418248450853],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(255,153,0,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[129.66376187212376,131.22303506735113],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(221,68,119,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[133.99261445496637,138.75075184288931],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(51,102,204,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[141.52033123050455,147.17304276062723],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(34,170,153,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[149.94262214824244,196.67650473486174],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(153,0,153,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[199.44608412247695,200.0027695793876],"y":[-1,-1],"text":"","type":"scatter","mode":"lines","line":{"width":11.338582677165356,"color":"rgba(0,153,198,1)","dash":"solid"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[180],"y":[16.199999999999999],"text":"values above this threshold are capped","hovertext":"","textfont":{"size":7.559055118110237,"color":"rgba(190,190,190,1)"},"type":"scatter","mode":"text","hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[0,210],"y":[16,16],"text":"","type":"scatter","mode":"lines","line":{"width":0.75590551181102372,"color":"rgba(190,190,190,1)","dash":"dash"},"hoveron":"points","showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null}],"layout":{"margin":{"t":29.2835201328352,"r":6.6417600664176026,"b":55.878166138820184,"l":61.037775010377771},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0,210],"tickmode":"array","ticktext":["GO:MF","GO:CC","GO:BP","KEGG","REAC","MIRNA","HP","WP"],"tickvals":[17.596886321355132,41.562056762389666,88.412261683288804,130.44339846973745,136.37168314892784,144.3466869955659,173.30956344155209,199.72442685093228],"categoryorder":"array","categoryarray":["GO:MF","GO:CC","GO:BP","KEGG","REAC","MIRNA","HP","WP"],"nticks":null,"ticks":"","tickcolor":null,"ticklen":3.6529680365296811,"tickwidth":0,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":10.62681610626816},"tickangle":-45,"showline":true,"linecolor":"rgba(190,190,190,1)","linewidth":0.132835201328352,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":"","hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[-1,18.5],"tickmode":"array","ticktext":["0","2","4","6","8","10","12","14",">16"],"tickvals":[0,2,4,6,8,10,12,14,16],"categoryorder":"array","categoryarray":["0","2","4","6","8","10","12","14",">16"],"nticks":null,"ticks":"outside","tickcolor":"rgba(190,190,190,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":true,"linecolor":"rgba(190,190,190,1)","linewidth":0.66417600664176002,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":"","hoverformat":".2f"},"annotations":[{"text":"-log10(p-adj)","x":0,"y":0.5,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(0,0,0,1)","family":"","size":13.283520132835198},"xref":"paper","yref":"paper","textangle":-90,"xanchor":"right","yanchor":"center","annotationType":"axis","xshift":-37.393109173931094},{"text":"query_1","x":0.5,"y":1,"showarrow":false,"ax":0,"ay":0,"font":{"color":"rgba(169,169,169,1)","family":"","size":13.283520132835198},"xref":"paper","yref":"paper","textangle":-0,"xanchor":"center","yanchor":"bottom"}],"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1},{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":24.973017849730184,"yanchor":1,"ysizemode":"pixel"}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498}},"hovermode":"closest","barmode":"relative","dragmode":"zoom"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"7787ea70b1f":{"x":{},"y":{},"text":{},"colour":{},"size":{},"alpha":{},"type":"scatter"},"778622db7bb":{"x":{},"xend":{},"y":{},"yend":{}},"77857f3865a":{"x":{},"xend":{},"y":{},"yend":{}},"7785a95f69a":{"x":{},"xend":{},"y":{},"yend":{}},"7785392aecd":{"x":{},"xend":{},"y":{},"yend":{}},"7785ec29943":{"x":{},"xend":{},"y":{},"yend":{}},"77823412cc3":{"x":{},"xend":{},"y":{},"yend":{}},"778539d2494":{"x":{},"xend":{},"y":{},"yend":{}},"77812c64e32":{"x":{},"xend":{},"y":{},"yend":{}},"7782cbd1911":{"x":{},"y":{}},"7785c616262":{"yintercept":{}}},"cur_data":"7787ea70b1f","visdat":{"7787ea70b1f":["function (y) ","x"],"778622db7bb":["function (y) ","x"],"77857f3865a":["function (y) ","x"],"7785a95f69a":["function (y) ","x"],"7785392aecd":["function (y) ","x"],"7785ec29943":["function (y) ","x"],"77823412cc3":["function (y) ","x"],"778539d2494":["function (y) ","x"],"77812c64e32":["function (y) ","x"],"7782cbd1911":["function (y) ","x"],"7785c616262":["function (y) ","x"]},"highlight":{"on":"plotly_click","off":"plotly_doubleclick","persistent":false,"dynamic":false,"color":null,"selectize":false,"defaultValues":null,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0,"ctGroups":["SharedData9beb5615"]},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
<pre class="r"><code>publish_gosttable(gost_results, highlight_terms = gost_results$result[c(1:10),],
                  use_colors = TRUE, 
                  show_columns = c(&quot;source&quot;, &quot;term_name&quot;, &quot;term_size&quot;, &quot;intersection_size&quot;),
                  filename = NULL)</code></pre>
<pre><code>## The input &#39;highlight_terms&#39; is a data.frame. The column &#39;term_id&#39; will be used.</code></pre>
<p><img src="tertiary_analysis_Part3_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>Write <code>gprofiler2</code> results to csv</p>
<pre class="r"><code>go_results &lt;- gost_results$result

go_results$parents &lt;- as.character(gost_results$result$parents)</code></pre>
<pre class="r"><code>write.csv(go_results,
          file = &quot;./results/Transplant_vs_Naive/Transplant_vs_Naive_Gprofiler_padj_fc.csv&quot;,
          row.names = FALSE)</code></pre>
<div id="session-info" class="section level2">
<h2>session info</h2>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.3.3 (2024-02-29)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 22.04.4 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: Etc/UTC
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] RColorBrewer_1.1-3    ComplexHeatmap_2.18.0 lubridate_1.9.3      
##  [4] forcats_1.0.0         stringr_1.5.1         purrr_1.0.2          
##  [7] readr_2.1.5           tidyr_1.3.1           tibble_3.2.1         
## [10] tidyverse_2.0.0       dplyr_1.1.4           ggrepel_0.9.5        
## [13] ggplot2_3.5.0         magrittr_2.0.3        gprofiler2_0.2.3     
## 
## loaded via a namespace (and not attached):
##  [1] tidyselect_1.2.1    viridisLite_0.4.2   bitops_1.0-7       
##  [4] fastmap_1.1.1       lazyeval_0.2.2      RCurl_1.98-1.14    
##  [7] promises_1.3.0      digest_0.6.35       timechange_0.3.0   
## [10] mime_0.12           lifecycle_1.0.4     cluster_2.1.6      
## [13] compiler_4.3.3      rlang_1.1.3         sass_0.4.9         
## [16] tools_4.3.3         utf8_1.2.4          yaml_2.3.8         
## [19] data.table_1.15.4   knitr_1.46          htmlwidgets_1.6.4  
## [22] withr_3.0.0         BiocGenerics_0.48.1 stats4_4.3.3       
## [25] fansi_1.0.6         xtable_1.8-4        colorspace_2.1-0   
## [28] scales_1.3.0        iterators_1.0.14    cli_3.6.2          
## [31] rmarkdown_2.26      crayon_1.5.2        generics_0.1.3     
## [34] rstudioapi_0.16.0   httr_1.4.7          tzdb_0.4.0         
## [37] rjson_0.2.21        cachem_1.0.8        parallel_4.3.3     
## [40] matrixStats_1.3.0   vctrs_0.6.5         jsonlite_1.8.8     
## [43] IRanges_2.36.0      hms_1.1.3           GetoptLong_1.0.5   
## [46] S4Vectors_0.40.2    clue_0.3-65         crosstalk_1.2.1    
## [49] foreach_1.5.2       plotly_4.10.4       jquerylib_0.1.4    
## [52] glue_1.7.0          codetools_0.2-20    stringi_1.8.3      
## [55] gtable_0.3.5        shape_1.4.6.1       later_1.3.2        
## [58] munsell_0.5.1       pillar_1.9.0        htmltools_0.5.8.1  
## [61] circlize_0.4.16     R6_2.5.1            doParallel_1.0.17  
## [64] evaluate_0.23       shiny_1.8.1.1       highr_0.10         
## [67] png_0.1-8           httpuv_1.6.15       bslib_0.7.0        
## [70] Rcpp_1.0.12         gridExtra_2.3       xfun_0.43          
## [73] pkgconfig_2.0.3     GlobalOptions_0.1.2</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
